// @cdklabs/generative-ai-cdk-constructs/lib/cdk-lib/bedrock/data-sources/custom-transformation.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomTransformation = exports.TransformationStep = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
const aws_cdk_lib_1 = require("aws-cdk-lib");
const aws_iam_1 = require("aws-cdk-lib/aws-iam");
/**
 * Defines the step in the ingestion process where the custom transformation is applied.
 */
var TransformationStep;
(function (TransformationStep) {
    /**
     * Processes documents after they have been converted into chunks.
     * This allows for custom chunk-level metadata addition or custom post-chunking logic.
     */
    TransformationStep["POST_CHUNKING"] = "POST_CHUNKING";
})(TransformationStep || (exports.TransformationStep = TransformationStep = {}));
/**
 * Represents a custom transformation configuration for a data source ingestion.
 * @see https://docs.aws.amazon.com/bedrock/latest/userguide/kb-chunking-parsing.html#kb-custom-transformation
 */
class CustomTransformation {
    // ------------------------------------------------------
    // Lambda Transformation Strategy
    // ------------------------------------------------------
    /**
     * This feature allows you to use a Lambda function to inject your own logic
     * into the knowledge base ingestion process.
     * @see https://github.com/aws-samples/amazon-bedrock-samples/blob/main/knowledge-bases/features-examples/02-optimizing-accuracy-retrieved-results/advanced_chunking_options.ipynb
     */
    static lambda(props) {
        class LambdaCustomTransformation extends CustomTransformation {
            constructor() {
                super(...arguments);
                this.configuration = {
                    intermediateStorage: {
                        s3Location: {
                            uri: props.s3BucketUri,
                        },
                    },
                    transformations: [
                        {
                            stepToApply: TransformationStep.POST_CHUNKING,
                            // To uncomment when more steps are available
                            // stepToApply: props.stepToApply ?? TransformationStep.POST_CHUNKING,
                            transformationFunction: {
                                transformationLambdaConfiguration: {
                                    lambdaArn: props.lambdaFunction.functionArn,
                                },
                            },
                        },
                    ],
                };
            }
            generatePolicyStatements(scope) {
                return [
                    new aws_iam_1.PolicyStatement({
                        actions: ['lambda:InvokeFunction'],
                        resources: [`${props.lambdaFunction.functionArn}:*`],
                        conditions: {
                            StringEquals: {
                                'aws:ResourceAccount': aws_cdk_lib_1.Stack.of(scope).account,
                            },
                        },
                    }),
                ];
            }
        }
        return new LambdaCustomTransformation();
    }
}
exports.CustomTransformation = CustomTransformation;
_a = JSII_RTTI_SYMBOL_1;
CustomTransformation[_a] = { fqn: "@cdklabs/generative-ai-cdk-constructs.bedrock.CustomTransformation", version: "0.1.312" };
