// awslint/lib/rules/api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.apiLinter = void 0;
const construct_1 = require("./construct");
const core_types_1 = require("./core-types");
const linter_1 = require("../linter");
const CLASS_ON_PURPOSE_RULE = 'ref-via-interface';
const L2_INTERFACE_ON_PURPOSE_RULE = 'prefer-ref-interface';
// lint all constructs that are not L1 resources
exports.apiLinter = new linter_1.Linter(a => construct_1.ConstructReflection
    .findAllConstructs(a)
    .filter(c => !core_types_1.CoreTypes.isCfnResource(c.classType)));
exports.apiLinter.add({
    code: CLASS_ON_PURPOSE_RULE,
    message: 'API should use interface (preferably IxxxRef, otherwise Ixxx) and not the concrete class (%s). ' +
        `If this is intentional, add "${ignoreTag(CLASS_ON_PURPOSE_RULE)}" to element's jsdoc`,
    eval: e => {
        visitParameters(e.ctx.classType, {
            visitType: function (type, docs, scope) {
                // allow exclusion of this rule
                if (docHasIgnoreTag(docs, CLASS_ON_PURPOSE_RULE)) {
                    return;
                }
                // Assert that all types that reference classes are not resource constructs. Those
                // should be referenced via Interface.
                e.assert(!type.type
                    || !type.type.isClassType()
                    || !core_types_1.CoreTypes.isResourceClass(type.type)
                    // Make an exception for `constructs.Construct`.
                    || type.type.fqn === e.ctx.core.baseConstructClassFqn, scope, type.type?.fqn);
            },
        });
    },
});
exports.apiLinter.add({
    code: L2_INTERFACE_ON_PURPOSE_RULE,
    message: 'API should prefer to use the L1 reference interface (IxxxRef) and not the L2 interface (%s). ' +
        `If this is intentional, add "${ignoreTag(L2_INTERFACE_ON_PURPOSE_RULE)}" to element's jsdoc`,
    eval: e => {
        visitParameters(e.ctx.classType, {
            visitType: function (type, docs, scope) {
                // allow exclusion of this rule
                if (docHasIgnoreTag(docs, L2_INTERFACE_ON_PURPOSE_RULE)) {
                    return;
                }
                // Assert that all types that interfaces are not L2 interfaces. This is actually not checking
                // that they ARE Ref interfaces, just that they aren't L2 interfaces (could also be service-
                // specific non-resource interfaces, and those are fine too)
                e.assert(!type.type
                    || !type.type.isInterfaceType()
                    || type.type.datatype
                    || !core_types_1.CoreTypes.isL2Interface(type.type), scope, type.type?.fqn);
            },
        });
    },
});
function ignoreTag(rule) {
    return `[disable-awslint:${rule}]`;
}
function docHasIgnoreTag(docs, rule) {
    const tag = ignoreTag(rule);
    return docs.summary.includes(tag) || docs.remarks.includes(tag);
}
/**
 * Visit all types in the API of a given type
 *
 * This visits method parameters and members of structs of method parameters.
 */
function visitParameters(root, visitor) {
    const visited = new Set();
    if (root.isClassType()) {
        assertClass(root);
    }
    else if (root.isInterfaceType()) {
        assertInterface(root);
    }
    else {
        throw new Error(`invalid type: ${root.toString()}`);
    }
    function assertClass(type) {
        if (visited.has(type.fqn)) {
            return;
        }
        visited.add(type.fqn);
        for (const method of type.allMethods) {
            assertMethod(method);
        }
        if (type.initializer) {
            assertMethod(type.initializer);
        }
    }
    function assertDataType(type) {
        for (const property of type.allProperties) {
            assertProperty(property);
        }
    }
    function assertInterface(type) {
        if (visited.has(type.fqn)) {
            return;
        }
        visited.add(type.fqn);
        if (type.datatype) {
            assertDataType(type);
        }
        for (const method of type.allMethods) {
            assertMethod(method);
        }
    }
    function assertProperty(property) {
        if (property.protected) {
            return;
        }
        const site = property.overrides ? property.overrides : property.parentType;
        assertType(property.type, property.docs, `${site.fqn}.${property.name}`);
    }
    function assertMethod(method) {
        if (method.protected) {
            return;
        }
        const site = method.overrides ? method.overrides : method.parentType;
        const scope = `${site.fqn}.${method.name}`;
        let firstMethod = site.isClassType() || site.isInterfaceType()
            ? site.allMethods.find(m => m.name === method.name)
            : undefined;
        if (!firstMethod) {
            firstMethod = method;
        }
        for (const param of firstMethod.parameters) {
            assertType(param.type, param.docs, `${scope}.${param.name}`);
        }
        // note that we do not require that return values will use an interface
    }
    function assertType(type, docs, scope) {
        visitor.visitType(type, docs, scope);
        // And recurse
        if (type.arrayOfType) {
            return assertType(type.arrayOfType, docs, scope);
        }
        if (type.mapOfType) {
            return assertType(type.mapOfType, docs, scope);
        }
        if (type.unionOfTypes) {
            for (const t of type.unionOfTypes) {
                assertType(t, docs, scope);
            }
            return;
        }
        // interfaces are okay
        if (type.type && type.type.isInterfaceType()) {
            return assertInterface(type.type);
        }
    }
}
