// awslint/lib/rules/attributes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.attributesLinter = void 0;
const resource_1 = require("./resource");
const linter_1 = require("../linter");
exports.attributesLinter = new linter_1.Linter(a => {
    const result = new Array();
    for (const resource of resource_1.ResourceReflection.findAll(a)) {
        for (const attr of resource.attributes) {
            result.push(new AttributeReflection(resource, attr));
        }
    }
    return result;
});
class AttributeReflection {
    resource;
    attr;
    fqn;
    constructor(resource, attr) {
        this.resource = resource;
        this.attr = attr;
        this.fqn = resource.fqn + '.' + attr.property.name;
    }
}
exports.attributesLinter.add({
    code: 'attribute-readonly',
    message: 'attribute property must be readonly',
    eval: e => {
        e.assert(e.ctx.attr.property.immutable, e.ctx.fqn);
    },
});
exports.attributesLinter.add({
    code: 'attribute-tag',
    message: 'attribute properties must have an "@attribute" doctag on: ',
    eval: e => {
        const tag = e.ctx.attr.property.docs.customTag('attribute');
        e.assert(tag, e.ctx.fqn, `${e.ctx.attr.property.parentType.fqn}.${e.ctx.attr.property.name}`);
    },
});
