// awslint/lib/rules/cfn-resource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CfnResourceReflection = exports.cfnResourceLinter = void 0;
const core_types_1 = require("./core-types");
const resource_1 = require("./resource");
const case_1 = require("../case");
const linter_1 = require("../linter");
const cfnResourceTagName = 'cloudformationResource';
// this linter verifies that we have L2 coverage. it finds all "Cfn" classes and verifies
// that we have a corresponding L1 class for it that's identified as a resource.
exports.cfnResourceLinter = new linter_1.Linter(a => CfnResourceReflection.findAll(a));
// Cache L1 constructs per type system.
const l1ConstructCache = new Map();
function cacheL1ConstructsForTypeSystem(sys) {
    if (!l1ConstructCache.has(sys)) {
        l1ConstructCache.set(sys, new Map());
        for (const cls of sys.classes) {
            const cfnResourceTag = cls.docs.customTag(cfnResourceTagName);
            if (cfnResourceTag) {
                l1ConstructCache.get(sys)?.set(cfnResourceTag?.toLocaleLowerCase(), cls);
            }
        }
    }
}
exports.cfnResourceLinter.add({
    code: 'resource-class',
    message: 'every resource must have a resource class (L2), add \'@resource %s\' to its docstring',
    warning: true,
    eval: e => {
        const l2 = resource_1.ResourceReflection.findAll(e.ctx.classType.assembly).find(r => r.cfn.fullname === e.ctx.fullname);
        e.assert(l2, e.ctx.fullname, e.ctx.fullname);
    },
});
class CfnResourceReflection {
    /**
     * Finds a Cfn resource class by full CloudFormation resource name (e.g. `AWS::S3::Bucket`)
     * @param fullName first two components are case-insensitive (e.g. `aws::s3::Bucket` is equivalent to `Aws::S3::Bucket`)
     */
    static findByName(sys, fullName) {
        if (!l1ConstructCache.has(sys)) {
            cacheL1ConstructsForTypeSystem(sys);
        }
        const cls = l1ConstructCache.get(sys)?.get(fullName.toLowerCase());
        if (cls) {
            return new CfnResourceReflection(cls);
        }
        return undefined;
    }
    /**
     * Returns all CFN resource classes within an assembly.
     */
    static findAll(assembly) {
        return assembly.allClasses
            .filter(c => core_types_1.CoreTypes.isCfnResource(c))
            .map(c => new CfnResourceReflection(c));
    }
    classType;
    fullname; // AWS::S3::Bucket
    namespace; // AWS::S3
    basename; // Bucket
    attributeNames; // (normalized) bucketArn, bucketName, queueUrl
    doc; // link to CloudFormation docs
    constructor(cls) {
        this.classType = cls;
        this.basename = cls.name.slice('Cfn'.length);
        const fullname = cls.docs.customTag('cloudformationResource');
        if (!fullname) {
            throw new Error(`Unable to extract CloudFormation resource name from initializer documentation of ${cls}`);
        }
        this.fullname = fullname;
        this.namespace = fullname.split('::').slice(0, 2).join('::');
        this.attributeNames = cls.ownProperties
            .filter(p => (p.docs.docs.custom || {}).cloudformationAttribute)
            .map(p => p.docs.customTag('cloudformationAttribute') || '<error>')
            .map(p => this.attributePropertyNameFromCfnName(p));
        this.doc = cls.docs.docs.see || '';
    }
    attributePropertyNameFromCfnName(name) {
        // special case (someone was smart), special case copied from spec2cdk
        if (this.basename === 'SecurityGroup' && name === 'GroupId') {
            return 'Id';
        }
        return (0, case_1.pascalize)(name);
    }
}
exports.CfnResourceReflection = CfnResourceReflection;
