// awslint/lib/rules/cloudwatch-events.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventsReflection = exports.eventsLinter = void 0;
const construct_1 = require("./construct");
const core_types_1 = require("./core-types");
const linter_1 = require("../linter");
exports.eventsLinter = new linter_1.Linter(assembly => assembly.classes
    .filter(t => core_types_1.CoreTypes.isConstructClass(t))
    .map(construct => new EventsReflection(construct)));
class EventsReflection extends construct_1.ConstructReflection {
    get directEventMethods() {
        return this.classType.allMethods.filter(isDirectEventMethod);
    }
    get cloudTrailEventMethods() {
        return this.classType.allMethods.filter(isCloudTrailEventMethod);
    }
}
exports.EventsReflection = EventsReflection;
const ON_EVENT_OPTIONS_FQN = 'aws-cdk-lib.aws_events.OnEventOptions';
const EVENT_RULE_FQN = 'aws-cdk-lib.aws_events.Rule';
exports.eventsLinter.add({
    code: 'events-in-interface',
    message: '\'onXxx()\' methods should also be defined on construct interface',
    eval: e => {
        for (const method of e.ctx.directEventMethods.concat(e.ctx.cloudTrailEventMethods)) {
            e.assert(!e.ctx.interfaceType || e.ctx.interfaceType.allMethods.filter(m => !m.protected).some(m => m.name === method.name), `${e.ctx.fqn}.${method.name}`);
        }
    },
});
exports.eventsLinter.add({
    code: 'events-generic',
    message: 'if there are specific \'onXxx()\' methods, there should also be a generic \'onEvent()\' method',
    eval: e => {
        e.assert(e.ctx.directEventMethods.length === 0 || e.ctx.classType.allMethods.some(m => m.name === 'onEvent'), e.ctx.fqn);
    },
});
exports.eventsLinter.add({
    code: 'events-generic-cloudtrail',
    message: 'if there are specific \'onCloudTrailXxx()\' methods, there should also be a generic \'onCloudTrailEvent()\' method',
    eval: e => {
        e.assert(e.ctx.cloudTrailEventMethods.length === 0 || e.ctx.classType.allMethods.some(m => m.name === 'onCloudTrailEvent'), e.ctx.fqn);
    },
});
exports.eventsLinter.add({
    code: 'events-method-signature',
    message: 'all \'onXxx()\' methods should have the CloudWatch Events signature (id: string, options: events.OnEventOptions = {}) => events.Rule',
    eval: e => {
        for (const method of e.ctx.directEventMethods) {
            // give slack to protected methods like "onSynthesize", "onPrepare", ...
            if (method.protected) {
                continue;
            }
            e.assertSignature(method, {
                parameters: [
                    { type: 'string' },
                    { type: ON_EVENT_OPTIONS_FQN, subtypeAllowed: true, optional: true },
                ],
                returns: EVENT_RULE_FQN,
            });
        }
    },
});
function isDirectEventMethod(m) {
    return !m.protected && m.name.startsWith('on') && !m.name.startsWith('onCloudTrail');
}
function isCloudTrailEventMethod(m) {
    return m.name.startsWith('onCloudTrail');
}
