// awslint/lib/rules/core-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CoreTypes = void 0;
const util_1 = require("./util");
const CORE_MODULE = 'aws-cdk-lib';
var CoreTypesFqn;
(function (CoreTypesFqn) {
    CoreTypesFqn["CfnResource"] = "aws-cdk-lib.CfnResource";
    CoreTypesFqn["Resource"] = "aws-cdk-lib.Resource";
    CoreTypesFqn["ResourceInterface"] = "aws-cdk-lib.IResource";
    CoreTypesFqn["ResolvableInterface"] = "aws-cdk-lib.IResolvable";
    CoreTypesFqn["PhysicalName"] = "aws-cdk-lib.PhysicalName";
    CoreTypesFqn["Construct"] = "constructs.Construct";
    CoreTypesFqn["ConstructInterface"] = "constructs.IConstruct";
})(CoreTypesFqn || (CoreTypesFqn = {}));
class CoreTypes {
    /**
     * @returns true if assembly has the Core module
     */
    static hasCoreModule(assembly) {
        return (!assembly.system.assemblies.find(a => a.name === CORE_MODULE));
    }
    /**
     * @returns true if `classType` represents an L1 Cfn Resource
     */
    static isCfnResource(c) {
        if (!c.system.includesAssembly(CORE_MODULE)) {
            return false;
        }
        // skip CfnResource itself
        if (c.fqn === CoreTypesFqn.CfnResource) {
            return false;
        }
        if (!this.isConstructClass(c)) {
            return false;
        }
        const cfnResourceClass = c.system.findFqn(CoreTypesFqn.CfnResource);
        if (!c.extends(cfnResourceClass)) {
            return false;
        }
        return c.name.startsWith('Cfn');
    }
    /**
     * @returns true if `classType` represents a Construct
     */
    static isConstructClass(c) {
        if (!c.system.includesAssembly(CORE_MODULE)) {
            return false;
        }
        if (!c.isClassType()) {
            return false;
        }
        if (c.abstract) {
            return false;
        }
        return c.extends(c.system.findFqn(CoreTypesFqn.Construct));
    }
    /**
     * @returns true if `classType` represents an AWS resource (i.e. extends `cdk.Resource`).
     */
    static isResourceClass(classType) {
        const baseResource = classType.system.findClass(CoreTypesFqn.Resource);
        return classType.extends(baseResource) || (0, util_1.getDocTag)(classType, 'resource');
    }
    /**
     * @returns true if `interfaceType` looks like an L2 interface (`IBucket`)
     */
    static isL2Interface(interfaceType) {
        // We determine this by it being a behavioral interface, and it inheriting from `IResource`.
        const baseInterface = interfaceType.system.findInterface(CoreTypesFqn.ResourceInterface);
        return !interfaceType.datatype && interfaceExtends(interfaceType, baseInterface);
    }
    /**
     * @returns true if `interfaceType` looks like an L1 Ref interface (`IBucketRef)`
     */
    static isL1RefInterface(interfaceType) {
        // We determine this by it being a behavioral interface, and it living inside the `aws-cdk-lib.interfaces` namespace
        // with a name ending in `Ref`.
        return !interfaceType.datatype && interfaceType.fqn.startsWith('aws-cdk-lib.interfaces.') && interfaceType.name.endsWith('Ref');
    }
    /**
     * Return true if the nesting parent of the given interface is a CFN class
     */
    static isCfnNestedType(interfaceType) {
        return interfaceType.nestingParent && CoreTypes.isCfnType(interfaceType.nestingParent);
    }
    /**
     * Return true if the given interface type is a CFN class, prop type or interface
     */
    static isCfnType(interfaceType) {
        // aws_service.CfnTheResource
        if (interfaceType.name.startsWith('Cfn')) {
            return true;
        }
        // aws_service.ITheResourceRf
        if (/^I\w+Ref/.test(interfaceType.name)) {
            return true;
        }
        if (interfaceType.namespace) {
            if (interfaceType.namespace.startsWith('Cfn')) {
                return true;
            }
            const namespaceParts = interfaceType.namespace.split('.');
            // aws_service.CfnTheResource.SubType
            if (namespaceParts.at(1)?.startsWith('Cfn')) {
                return true;
            }
            // aws_service.mixins.CfnTheResource.SubType
            if (namespaceParts.at(1) === 'mixins' && namespaceParts.at(2)?.startsWith('Cfn')) {
                return true;
            }
        }
        return false;
    }
    /**
     * @returns `classType` for the core type Construct
     * @deprecated - use `baseConstructClass()`
     */
    get constructClass() {
        return this.baseConstructClass;
    }
    /**
     * @returns `classType` for the core type Construct
     */
    get baseConstructClass() {
        return this.sys.findClass(this.baseConstructClassFqn);
    }
    /**
     * @returns `classType` for the core type Construct
     */
    get baseConstructClassFqn() {
        return CoreTypesFqn.Construct;
    }
    /**
     * @returns `interfacetype` for the core type Construct
     * @deprecated - use `baseConstructInterface()`
     */
    get constructInterface() {
        return this.baseConstructInterface;
    }
    /**
     * @returns `interfacetype` for the core type Construct
     */
    get baseConstructInterface() {
        return this.sys.findInterface(this.baseConstructInterfaceFqn);
    }
    /**
     * @returns fqn for for the core Construct interface
     */
    get baseConstructInterfaceFqn() {
        return CoreTypesFqn.ConstructInterface;
    }
    /**
     * @returns `classType` for the core type Resource
     */
    get resourceClass() {
        return this.sys.findClass(this.resourceClassFqn);
    }
    /**
     * @returns fqn for the core type Resource
     */
    get resourceClassFqn() {
        return CoreTypesFqn.Resource;
    }
    /**
     * @returns fqn for the core Resource interface
     */
    get resourceInterface() {
        return this.sys.findInterface(this.resourceInterfaceFqn);
    }
    /**
     * @returns `interfaceType` for the core type Resource
     */
    get resourceInterfaceFqn() {
        return CoreTypesFqn.ResourceInterface;
    }
    /**
     * @returns `classType` for the core type Token
     */
    get tokenInterface() {
        return this.sys.findInterface(this.tokenInterfaceFqn);
    }
    /**
     * @returns fqn for the core type Token
     */
    get tokenInterfaceFqn() {
        return CoreTypesFqn.ResolvableInterface;
    }
    get physicalNameClass() {
        return this.sys.findClass(CoreTypesFqn.PhysicalName);
    }
    sys;
    constructor(sys) {
        this.sys = sys;
        if (!sys.includesAssembly(CORE_MODULE)) {
            // disable-all-checks
            return;
        }
    }
}
exports.CoreTypes = CoreTypes;
function interfaceExtends(interfaceType, baseInterface) {
    return interfaceType.getInterfaces(true).some(i => i.fqn === baseInterface.fqn);
}
