// awslint/lib/rules/grant-methods.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.grantsMethodsLinter = void 0;
const linter_1 = require("../linter");
const EXCLUDE_ANNOTATION_NO_GRANTS = '[disable-awslint:no-grants]';
exports.grantsMethodsLinter = new linter_1.Linter(assembly => assembly.allClasses);
exports.grantsMethodsLinter.add({
    code: 'no-grants',
    message: 'L2 constructs are not allowed to have new "grantXxx()" methods. Create or re-use a companion <Contruct>Grants class instead.',
    eval: e => {
        Object.values(e.ctx.getMethods())
            .filter(m => !m.static)
            .filter(m => !m.abstract)
            .filter(m => m.name.startsWith('grant') && m.name !== 'grantOnKey')
            .filter(m => m.parameters.length > 0 && m.parameters[0].type.fqn === 'aws-cdk-lib.aws_iam.IGrantable')
            .forEach((grant) => {
            e.assert(isExempted(grant), `${e.ctx.fqn}.${grant.name}`);
        });
    },
});
function isExempted(method) {
    const docs = method.docs;
    return [docs.summary, docs.remarks].some(doc => doc.includes(EXCLUDE_ANNOTATION_NO_GRANTS));
}
