// awslint/lib/rules/integrations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.integrationLinter = void 0;
const util_1 = require("./util");
const linter_1 = require("../linter");
exports.integrationLinter = new linter_1.Linter(assembly => assembly.allInterfaces
    .filter(IntegrationReflection.isIntegrationInterface)
    .map(construct => new IntegrationReflection(construct)));
const BIND_METHOD_NAME = 'bind';
class IntegrationReflection {
    integrationInterface;
    static isIntegrationInterface(x) {
        return x.allMethods.some(m => m.name === BIND_METHOD_NAME);
    }
    constructor(integrationInterface) {
        this.integrationInterface = integrationInterface;
    }
    get bindMethod() {
        return this.integrationInterface.allMethods.find(m => m.name === BIND_METHOD_NAME);
    }
}
exports.integrationLinter.add({
    code: 'integ-return-type',
    message: '\'bind(...)\' should return a type named \'XxxConfig\'',
    eval: e => {
        const returnsFqn = e.ctx.bindMethod.returns.type.fqn;
        e.assert(returnsFqn && returnsFqn.endsWith('Config'), (0, util_1.memberFqn)(e.ctx.bindMethod));
    },
});
