// @aws-cdk/aws-bedrock-alpha/bedrock/guardrails/guardrails.js
"use strict";
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Guardrail = exports.GuardrailBase = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const fs = require("fs");
const aws_cdk_lib_1 = require("aws-cdk-lib");
const bedrock = require("aws-cdk-lib/aws-bedrock");
const aws_cloudwatch_1 = require("aws-cdk-lib/aws-cloudwatch");
const iam = require("aws-cdk-lib/aws-iam");
const aws_kms_1 = require("aws-cdk-lib/aws-kms");
const helpers_internal_1 = require("aws-cdk-lib/core/lib/helpers-internal");
const metadata_resource_1 = require("aws-cdk-lib/core/lib/metadata-resource");
const prop_injectable_1 = require("aws-cdk-lib/core/lib/prop-injectable");
// Internal Libs
const filters = require("./guardrail-filters");
const guardrail_version_1 = require("./guardrail-version");
/**
 * Abstract base class for a Guardrail.
 * Contains methods and attributes valid for Guardrails either created with CDK or imported.
 */
class GuardrailBase extends aws_cdk_lib_1.Resource {
    static [JSII_RTTI_SYMBOL_1] = { fqn: "@aws-cdk/aws-bedrock-alpha.GuardrailBase", version: "2.235.0-alpha.0" };
    /**
     * Return the given named metric for all guardrails.
     *
     * By default, the metric will be calculated as a sum over a period of 5 minutes.
     * You can customize this by using the `statistic` and `period` properties.
     */
    static metricAll(metricName, props) {
        return new aws_cloudwatch_1.Metric({
            namespace: 'AWS/Bedrock/Guardrails',
            dimensionsMap: { Operation: 'ApplyGuardrail' },
            metricName,
            ...props,
        });
    }
    /**
     * Return the invocations metric for all guardrails.
     */
    static metricAllInvocations(props) {
        return this.metricAll('Invocations', props);
    }
    /**
     * Return the text unit count metric for all guardrails.
     */
    static metricAllTextUnitCount(props) {
        return this.metricAll('TextUnitCount', props);
    }
    /**
     * Return the invocations intervened metric for all guardrails.
     */
    static metricAllInvocationsIntervened(props) {
        return this.metricAll('InvocationsIntervened', props);
    }
    /**
     * Return the invocation latency metric for all guardrails.
     */
    static metricAllInvocationLatency(props) {
        return this.metricAll('InvocationLatency', props);
    }
    _version = 'DRAFT';
    /**
     * The version of the guardrail.
     * @attribute
     */
    get guardrailVersion() {
        return this._version;
    }
    updateVersion(version) {
        this._version = version;
    }
    /**
     * Grant the given principal identity permissions to perform actions on this guardrail.
     * [disable-awslint:no-grants]
     */
    grant(grantee, ...actions) {
        return iam.Grant.addToPrincipal({
            grantee,
            actions,
            resourceArns: [this.guardrailArn],
        });
    }
    /**
     * Grant the given identity permissions to apply the guardrail.
     * [disable-awslint:no-grants]
     */
    grantApply(grantee) {
        const baseGrant = this.grant(grantee, 'bedrock:ApplyGuardrail');
        if (this.kmsKey) {
            // If KMS key exists, create encryption grant and combine with base grant
            const kmsGrant = this.kmsKey.grantEncryptDecrypt(grantee);
            return kmsGrant.combine(baseGrant);
        }
        else {
            // If no KMS key exists, return only the base grant
            return baseGrant;
        }
    }
    /**
     * Return the given named metric for this guardrail.
     *
     * By default, the metric will be calculated as a sum over a period of 5 minutes.
     * You can customize this by using the `statistic` and `period` properties.
     */
    metric(metricName, props) {
        const metricProps = {
            namespace: 'AWS/Bedrock/Guardrails',
            metricName,
            dimensionsMap: { GuardrailArn: this.guardrailArn, GuardrailVersion: this.guardrailVersion },
            ...props,
        };
        return this.configureMetric(metricProps);
    }
    /**
     * Return the invocations metric for this guardrail.
     */
    metricInvocations(props) {
        return this.metric('Invocations', props);
    }
    /**
     * Return the invocation latency metric for this guardrail.
     */
    metricInvocationLatency(props) {
        return this.metric('InvocationLatency', props);
    }
    /**
     * Return the invocation client errors metric for this guardrail.
     */
    metricInvocationClientErrors(props) {
        return this.metric('InvocationClientErrors', props);
    }
    /**
     * Return the invocation server errors metric for this guardrail.
     */
    metricInvocationServerErrors(props) {
        return this.metric('InvocationServerErrors', props);
    }
    /**
     * Return the invocation throttles metric for this guardrail.
     */
    metricInvocationThrottles(props) {
        return this.metric('InvocationThrottles', props);
    }
    /**
     * Return the text unit count metric for this guardrail.
     */
    metricTextUnitCount(props) {
        return this.metric('TextUnitCount', props);
    }
    /**
     * Return the invocations intervened metric for this guardrail.
     */
    metricInvocationsIntervened(props) {
        return this.metric('InvocationsIntervened', props);
    }
    configureMetric(props) {
        return new aws_cloudwatch_1.Metric({
            ...props,
            region: props?.region ?? this.stack.region,
            account: props?.account ?? this.stack.account,
        });
    }
}
exports.GuardrailBase = GuardrailBase;
/******************************************************************************
 *                        NEW CONSTRUCT DEFINITION
 *****************************************************************************/
/**
 * Class to create a Guardrail with CDK.
 * @cloudformationResource AWS::Bedrock::Guardrail
 */
let Guardrail = (() => {
    let _classDecorators = [prop_injectable_1.propertyInjectable];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    let _classSuper = GuardrailBase;
    let _instanceExtraInitializers = [];
    let _addContentFilter_decorators;
    let _addPIIFilter_decorators;
    let _addRegexFilter_decorators;
    let _addDeniedTopicFilter_decorators;
    let _addContextualGroundingFilter_decorators;
    let _addWordFilter_decorators;
    let _addWordFilterFromFile_decorators;
    let _addManagedWordListFilter_decorators;
    let _createVersion_decorators;
    var Guardrail = class extends _classSuper {
        static { _classThis = this; }
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _addContentFilter_decorators = [(0, metadata_resource_1.MethodMetadata)()];
            _addPIIFilter_decorators = [(0, metadata_resource_1.MethodMetadata)()];
            _addRegexFilter_decorators = [(0, metadata_resource_1.MethodMetadata)()];
            _addDeniedTopicFilter_decorators = [(0, metadata_resource_1.MethodMetadata)()];
            _addContextualGroundingFilter_decorators = [(0, metadata_resource_1.MethodMetadata)()];
            _addWordFilter_decorators = [(0, metadata_resource_1.MethodMetadata)()];
            _addWordFilterFromFile_decorators = [(0, metadata_resource_1.MethodMetadata)()];
            _addManagedWordListFilter_decorators = [(0, metadata_resource_1.MethodMetadata)()];
            _createVersion_decorators = [(0, metadata_resource_1.MethodMetadata)()];
            __esDecorate(this, null, _addContentFilter_decorators, { kind: "method", name: "addContentFilter", static: false, private: false, access: { has: obj => "addContentFilter" in obj, get: obj => obj.addContentFilter }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(this, null, _addPIIFilter_decorators, { kind: "method", name: "addPIIFilter", static: false, private: false, access: { has: obj => "addPIIFilter" in obj, get: obj => obj.addPIIFilter }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(this, null, _addRegexFilter_decorators, { kind: "method", name: "addRegexFilter", static: false, private: false, access: { has: obj => "addRegexFilter" in obj, get: obj => obj.addRegexFilter }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(this, null, _addDeniedTopicFilter_decorators, { kind: "method", name: "addDeniedTopicFilter", static: false, private: false, access: { has: obj => "addDeniedTopicFilter" in obj, get: obj => obj.addDeniedTopicFilter }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(this, null, _addContextualGroundingFilter_decorators, { kind: "method", name: "addContextualGroundingFilter", static: false, private: false, access: { has: obj => "addContextualGroundingFilter" in obj, get: obj => obj.addContextualGroundingFilter }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(this, null, _addWordFilter_decorators, { kind: "method", name: "addWordFilter", static: false, private: false, access: { has: obj => "addWordFilter" in obj, get: obj => obj.addWordFilter }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(this, null, _addWordFilterFromFile_decorators, { kind: "method", name: "addWordFilterFromFile", static: false, private: false, access: { has: obj => "addWordFilterFromFile" in obj, get: obj => obj.addWordFilterFromFile }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(this, null, _addManagedWordListFilter_decorators, { kind: "method", name: "addManagedWordListFilter", static: false, private: false, access: { has: obj => "addManagedWordListFilter" in obj, get: obj => obj.addManagedWordListFilter }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(this, null, _createVersion_decorators, { kind: "method", name: "createVersion", static: false, private: false, access: { has: obj => "createVersion" in obj, get: obj => obj.createVersion }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
            Guardrail = _classThis = _classDescriptor.value;
            if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        }
        static [JSII_RTTI_SYMBOL_1] = { fqn: "@aws-cdk/aws-bedrock-alpha.Guardrail", version: "2.235.0-alpha.0" };
        /** Uniquely identifies this class. */
        static PROPERTY_INJECTION_ID = '@aws-cdk.aws-bedrock-alpha.Guardrail';
        /**
         * Import a guardrail given its attributes
         */
        static fromGuardrailAttributes(scope, id, attrs) {
            try {
                jsiiDeprecationWarnings._aws_cdk_aws_bedrock_alpha_GuardrailAttributes(attrs);
            }
            catch (error) {
                if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                    Error.captureStackTrace(error, this.fromGuardrailAttributes);
                }
                throw error;
            }
            class Import extends GuardrailBase {
                guardrailArn = attrs.guardrailArn;
                guardrailId = aws_cdk_lib_1.Arn.split(attrs.guardrailArn, aws_cdk_lib_1.ArnFormat.SLASH_RESOURCE_NAME).resourceName;
                kmsKey = attrs.kmsKey;
                lastUpdated = undefined;
                constructor() {
                    super(scope, id);
                    this.updateVersion(attrs.guardrailVersion ?? 'DRAFT');
                }
            }
            return new Import();
        }
        /**
         * Import a low-level L1 Cfn Guardrail
         */
        static fromCfnGuardrail(cfnGuardrail) {
            return new (class extends GuardrailBase {
                guardrailArn = cfnGuardrail.attrGuardrailArn;
                guardrailId = cfnGuardrail.attrGuardrailId;
                kmsKey = cfnGuardrail.kmsKeyArn
                    ? aws_kms_1.Key.fromKeyArn(this, '@FromCfnGuardrailKey', cfnGuardrail.kmsKeyArn)
                    : undefined;
                lastUpdated = cfnGuardrail.attrUpdatedAt;
                constructor() {
                    super(cfnGuardrail, '@FromCfnGuardrail');
                    this.updateVersion(cfnGuardrail.attrVersion);
                }
            })();
        }
        /**
         * The ARN of the guardrail.
         * @attribute
         */
        guardrailArn = __runInitializers(this, _instanceExtraInitializers);
        /**
         * The ID of the guardrail.
         * @attribute
         */
        guardrailId;
        /**
         * The name of the guardrail.
         */
        name;
        /**
         * The KMS key used to encrypt data.
         *
         * @default undefined - "Data is encrypted by default with a key that AWS owns and manages for you"
         */
        kmsKey;
        /**
         * The content filters applied by the guardrail.
         */
        contentFilters;
        /**
         * The PII filters applied by the guardrail.
         */
        piiFilters;
        /**
         * The regex filters applied by the guardrail.
         */
        regexFilters;
        /**
         * The denied topic filters applied by the guardrail.
         */
        deniedTopics;
        /**
         * The contextual grounding filters applied by the guardrail.
         */
        contextualGroundingFilters;
        /**
         * The word filters applied by the guardrail.
         */
        wordFilters;
        /**
         * The managed word list filters applied by the guardrail.
         */
        managedWordListFilters;
        /**
         * When this guardrail was last updated
         */
        lastUpdated;
        /**
         * The computed hash of the guardrail properties.
         */
        hash;
        /**
         * The L1 representation of the guardrail
         */
        __resource;
        /**
         * The tier that your guardrail uses for denied topic filters.
         * @default filters.TierConfig.CLASSIC
         */
        topicsTierConfig;
        /**
         * The tier that your guardrail uses for content filters.
         * Consider using a tier that balances performance, accuracy, and compatibility with your existing generative AI workflows.
         * @default filters.TierConfig.CLASSIC
         */
        contentFiltersTierConfig;
        /**
         * The cross-region configuration for the guardrail.
         */
        crossRegionConfig;
        /**
         * The message to return when the guardrail blocks a prompt.
         * @default "Sorry, your query violates our usage policy."
         */
        blockedInputMessaging;
        /**
         * The message to return when the guardrail blocks a model response.
         * @default "Sorry, I am unable to answer your question because of our usage policy."
         */
        blockedOutputsMessaging;
        constructor(scope, id, props) {
            super(scope, id, {
                physicalName: props.guardrailName,
            });
            try {
                jsiiDeprecationWarnings._aws_cdk_aws_bedrock_alpha_GuardrailProps(props);
            }
            catch (error) {
                if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                    Error.captureStackTrace(error, Guardrail);
                }
                throw error;
            }
            // Enhanced CDK Analytics Telemetry
            (0, metadata_resource_1.addConstructMetadata)(this, props);
            // ------------------------------------------------------
            // Set properties or defaults
            // ------------------------------------------------------
            this.name = this.physicalName;
            this.contentFilters = props.contentFilters ?? [];
            this.piiFilters = props.piiFilters ?? [];
            this.regexFilters = props.regexFilters ?? [];
            this.deniedTopics = props.deniedTopics ?? [];
            this.contextualGroundingFilters = props.contextualGroundingFilters ?? [];
            this.wordFilters = props.wordFilters ?? [];
            this.managedWordListFilters = props.managedWordListFilters ?? [];
            this.topicsTierConfig = props.topicsTierConfig ?? filters.TierConfig.CLASSIC;
            this.contentFiltersTierConfig = props.contentFiltersTierConfig ?? filters.TierConfig.CLASSIC;
            this.crossRegionConfig = props.crossRegionConfig;
            this.blockedInputMessaging = props.blockedInputMessaging ?? 'Sorry, your query violates our usage policy.';
            this.blockedOutputsMessaging = props.blockedOutputsMessaging ?? 'Sorry, I am unable to answer your question because of our usage policy.';
            // ------------------------------------------------------
            // Validate all filter arrays
            // ------------------------------------------------------
            this.validateContentFilters(this.contentFilters);
            this.validatePiiFilters(this.piiFilters);
            this.validateRegexFilters(this.regexFilters);
            this.validateDeniedTopics(this.deniedTopics);
            this.validateContextualGroundingFilters(this.contextualGroundingFilters);
            this.validateWordFilters(this.wordFilters);
            this.validateManagedWordListFilters(this.managedWordListFilters);
            // ------------------------------------------------------
            // Validate messaging properties
            // ------------------------------------------------------
            this.validateMessagingProperty(this.blockedInputMessaging, 'blockedInputMessaging');
            this.validateMessagingProperty(this.blockedOutputsMessaging, 'blockedOutputsMessaging');
            // ------------------------------------------------------
            // Validate tier configuration requirements
            // ------------------------------------------------------
            this.validateTierConfiguration(props);
            // ------------------------------------------------------
            // CFN Props - With Lazy support
            // ------------------------------------------------------
            let cfnProps = {
                name: props.guardrailName,
                description: props.description,
                kmsKeyArn: props.kmsKey?.keyArn,
                blockedInputMessaging: this.blockedInputMessaging,
                blockedOutputsMessaging: this.blockedOutputsMessaging,
                // Lazy props
                crossRegionConfig: this.generateCfnCrossRegionConfig(),
                contentPolicyConfig: this.generateCfnContentPolicyConfig(),
                contextualGroundingPolicyConfig: this.generateCfnContextualPolicyConfig(),
                topicPolicyConfig: this.generateCfnTopicPolicy(),
                wordPolicyConfig: this.generateCfnWordPolicyConfig(),
                sensitiveInformationPolicyConfig: this.generateCfnSensitiveInformationPolicyConfig(),
            };
            // Hash calculation useful for versioning of the guardrail
            this.hash = (0, helpers_internal_1.md5hash)(JSON.stringify(cfnProps));
            // ------------------------------------------------------
            // L1 Instantiation
            // ------------------------------------------------------
            this.__resource = new bedrock.CfnGuardrail(this, 'MyGuardrail', cfnProps);
            this.guardrailId = this.__resource.attrGuardrailId;
            this.guardrailArn = this.__resource.attrGuardrailArn;
            this.updateVersion(this.__resource.attrVersion);
            this.lastUpdated = this.__resource.attrUpdatedAt;
        }
        // ------------------------------------------------------
        // METHODS
        // ------------------------------------------------------
        /**
         * Adds a content filter to the guardrail.
         * @param filter The content filter to add.
         */
        addContentFilter(filter) {
            try {
                jsiiDeprecationWarnings._aws_cdk_aws_bedrock_alpha_ContentFilter(filter);
            }
            catch (error) {
                if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                    Error.captureStackTrace(error, this.addContentFilter);
                }
                throw error;
            }
            this.validateSingleContentFilter(filter);
            this.contentFilters.push(filter);
        }
        /**
         * Adds a PII filter to the guardrail.
         * @param filter The PII filter to add.
         */
        addPIIFilter(filter) {
            try {
                jsiiDeprecationWarnings._aws_cdk_aws_bedrock_alpha_PIIFilter(filter);
            }
            catch (error) {
                if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                    Error.captureStackTrace(error, this.addPIIFilter);
                }
                throw error;
            }
            this.validateSinglePiiFilter(filter);
            this.piiFilters.push(filter);
        }
        /**
         * Adds a regex filter to the guardrail.
         * @param filter The regex filter to add.
         */
        addRegexFilter(filter) {
            try {
                jsiiDeprecationWarnings._aws_cdk_aws_bedrock_alpha_RegexFilter(filter);
            }
            catch (error) {
                if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                    Error.captureStackTrace(error, this.addRegexFilter);
                }
                throw error;
            }
            this.validateSingleRegexFilter(filter);
            this.regexFilters.push(filter);
        }
        /**
         * Adds a denied topic filter to the guardrail.
         * @param filter The denied topic filter to add.
         */
        addDeniedTopicFilter(filter) {
            try {
                jsiiDeprecationWarnings._aws_cdk_aws_bedrock_alpha_Topic(filter);
            }
            catch (error) {
                if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                    Error.captureStackTrace(error, this.addDeniedTopicFilter);
                }
                throw error;
            }
            this.validateSingleDeniedTopic(filter);
            this.deniedTopics.push(filter);
        }
        /**
         * Adds a contextual grounding filter to the guardrail.
         * @param filter The contextual grounding filter to add.
         */
        addContextualGroundingFilter(filter) {
            try {
                jsiiDeprecationWarnings._aws_cdk_aws_bedrock_alpha_ContextualGroundingFilter(filter);
            }
            catch (error) {
                if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                    Error.captureStackTrace(error, this.addContextualGroundingFilter);
                }
                throw error;
            }
            this.validateSingleContextualGroundingFilter(filter);
            this.contextualGroundingFilters.push(filter);
        }
        /**
         * Adds a word filter to the guardrail.
         * @param filter The word filter to add.
         */
        addWordFilter(filter) {
            try {
                jsiiDeprecationWarnings._aws_cdk_aws_bedrock_alpha_WordFilter(filter);
            }
            catch (error) {
                if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                    Error.captureStackTrace(error, this.addWordFilter);
                }
                throw error;
            }
            this.validateSingleWordFilter(filter);
            this.wordFilters.push(filter);
        }
        /**
         * Adds a word filter to the guardrail.
         * @param filePath The location of the word filter file.
         */
        addWordFilterFromFile(filePath, inputAction, outputAction, inputEnabled, outputEnabled) {
            try {
                jsiiDeprecationWarnings._aws_cdk_aws_bedrock_alpha_GuardrailAction(inputAction);
                jsiiDeprecationWarnings._aws_cdk_aws_bedrock_alpha_GuardrailAction(outputAction);
            }
            catch (error) {
                if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                    Error.captureStackTrace(error, this.addWordFilterFromFile);
                }
                throw error;
            }
            const fileContents = fs.readFileSync(filePath, 'utf8');
            const words = fileContents.trim().split(',');
            for (const word of words)
                this.addWordFilter({ text: word, inputAction, outputAction, inputEnabled, outputEnabled });
        }
        /**
         * Adds a managed word list filter to the guardrail.
         * @param filter The managed word list filter to add.
         */
        addManagedWordListFilter(filter) {
            try {
                jsiiDeprecationWarnings._aws_cdk_aws_bedrock_alpha_ManagedWordFilter(filter);
            }
            catch (error) {
                if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                    Error.captureStackTrace(error, this.addManagedWordListFilter);
                }
                throw error;
            }
            this.validateSingleManagedWordListFilter(filter);
            this.managedWordListFilters.push(filter);
        }
        /**
         * Create a version for the guardrail.
         * @param description The description of the version.
         * @returns The guardrail version.
         */
        createVersion(description) {
            const cfnVersion = new guardrail_version_1.GuardrailVersion(this, `GuardrailVersion-${this.hash.slice(0, 16)}`, {
                description: description,
                guardrail: this,
            });
            this.updateVersion(cfnVersion.guardrailVersion);
            return this.guardrailVersion;
        }
        // ------------------------------------------------------
        // CFN Generators
        // ------------------------------------------------------
        /**
         * Returns the content filters applied to the guardrail. This method defers the computation
         * to synth time.
         */
        generateCfnContentPolicyConfig() {
            return aws_cdk_lib_1.Lazy.any({
                produce: () => {
                    if (this.contentFilters.length > 0) {
                        const contentPolicyConfig = {
                            filtersConfig: this.contentFilters,
                            ...(this.contentFiltersTierConfig && {
                                contentFiltersTierConfig: {
                                    tierName: this.contentFiltersTierConfig,
                                },
                            }),
                        };
                        return contentPolicyConfig;
                    }
                    else {
                        return undefined;
                    }
                },
            });
        }
        /**
         * Returns the topic filters applied to the guardrail. This method defers the computation
         * to synth time.
         */
        generateCfnTopicPolicy() {
            return aws_cdk_lib_1.Lazy.any({
                produce: () => {
                    if (this.deniedTopics.length > 0) {
                        const topicPolicyConfig = {
                            topicsConfig: this.deniedTopics.flatMap((topic) => {
                                return {
                                    definition: topic.definition,
                                    name: topic.name,
                                    examples: topic.examples,
                                    type: 'DENY',
                                    inputAction: topic.inputAction,
                                    inputEnabled: topic.inputEnabled,
                                    outputAction: topic.outputAction,
                                    outputEnabled: topic.outputEnabled,
                                };
                            }),
                            ...(this.topicsTierConfig && {
                                topicsTierConfig: {
                                    tierName: this.topicsTierConfig,
                                },
                            }),
                        };
                        return topicPolicyConfig;
                    }
                    else {
                        return undefined;
                    }
                },
            });
        }
        /**
         * Returns the contectual filters applied to the guardrail. This method defers the computation
         * to synth time.
         */
        generateCfnContextualPolicyConfig() {
            return aws_cdk_lib_1.Lazy.any({
                produce: () => {
                    if (this.contextualGroundingFilters.length > 0) {
                        return {
                            filtersConfig: this.contextualGroundingFilters.flatMap((filter) => {
                                return {
                                    type: filter.type,
                                    threshold: filter.threshold,
                                    action: filter.action,
                                    enabled: filter.enabled,
                                };
                            }),
                        };
                    }
                    else {
                        return undefined;
                    }
                },
            });
        }
        /**
         * Returns the word config applied to the guardrail. This method defers the computation
         * to synth time.
         */
        generateCfnWordPolicyConfig() {
            return aws_cdk_lib_1.Lazy.any({
                produce: () => {
                    if (this.wordFilters.length > 0 || this.managedWordListFilters.length > 0) {
                        return {
                            wordsConfig: this.generateCfnWordConfig(),
                            managedWordListsConfig: this.generateCfnManagedWordListsConfig(),
                        };
                    }
                    else {
                        return undefined;
                    }
                },
            });
        }
        /**
         * Returns the word filters applied to the guardrail. This method defers the computation
         * to synth time.
         */
        generateCfnWordConfig() {
            return aws_cdk_lib_1.Lazy.any({
                produce: () => {
                    return this.wordFilters.flatMap((word) => {
                        return {
                            text: word.text,
                            inputAction: word.inputAction,
                            inputEnabled: word.inputEnabled,
                            outputAction: word.outputAction,
                            outputEnabled: word.outputEnabled,
                        };
                    });
                },
            }, { omitEmptyArray: true });
        }
        /**
         * Returns the word filters applied to the guardrail. This method defers the computation
         * to synth time.
         */
        generateCfnManagedWordListsConfig() {
            return aws_cdk_lib_1.Lazy.any({
                produce: () => {
                    return this.managedWordListFilters.flatMap((filter) => {
                        return {
                            type: filter.type,
                            inputAction: filter.inputAction,
                            inputEnabled: filter.inputEnabled,
                            outputAction: filter.outputAction,
                            outputEnabled: filter.outputEnabled,
                        };
                    });
                },
            }, { omitEmptyArray: true });
        }
        /**
         * Returns the sensitive information config applied to the guardrail. This method defers the computation
         * to synth time.
         */
        generateCfnSensitiveInformationPolicyConfig() {
            return aws_cdk_lib_1.Lazy.any({
                produce: () => {
                    if (this.regexFilters.length > 0 || this.piiFilters.length > 0) {
                        return {
                            regexesConfig: this.generateCfnRegexesConfig(),
                            piiEntitiesConfig: this.generateCfnPiiEntitiesConfig(),
                        };
                    }
                    else {
                        return undefined;
                    }
                },
            }, { omitEmptyArray: true });
        }
        /**
         * Returns the regex filters applied to the guardrail. This method defers the computation
         * to synth time.
         */
        generateCfnRegexesConfig() {
            return aws_cdk_lib_1.Lazy.any({
                produce: () => {
                    return this.regexFilters.flatMap((regex) => {
                        return {
                            name: regex.name,
                            description: regex.description,
                            pattern: regex.pattern,
                            action: regex.action,
                            inputAction: regex.inputAction,
                            inputEnabled: regex.inputEnabled,
                            outputAction: regex.outputAction,
                            outputEnabled: regex.outputEnabled,
                        };
                    });
                },
            }, { omitEmptyArray: true });
        }
        /**
         * Returns the Pii filters applied to the guardrail. This method defers the computation
         * to synth time.
         */
        generateCfnPiiEntitiesConfig() {
            return aws_cdk_lib_1.Lazy.any({
                produce: () => {
                    return this.piiFilters.flatMap((filter) => {
                        return {
                            type: filter.type.value,
                            action: filter.action,
                            inputAction: filter.inputAction,
                            inputEnabled: filter.inputEnabled,
                            outputAction: filter.outputAction,
                            outputEnabled: filter.outputEnabled,
                        };
                    });
                },
            }, { omitEmptyArray: true });
        }
        /**
         * Returns the cross-region configuration for the guardrail. This method defers the computation
         * to synth time.
         */
        generateCfnCrossRegionConfig() {
            return aws_cdk_lib_1.Lazy.any({
                produce: () => {
                    if (this.crossRegionConfig) {
                        return {
                            guardrailProfileArn: this.crossRegionConfig.guardrailProfileArn,
                        };
                    }
                    else {
                        return undefined;
                    }
                },
            });
        }
        /**
         * Validates a RegexFilter object and applies default values for optional properties.
         * @param filter The regex filter to validate.
         * @param index Optional index for error messages when validating arrays.
         */
        validateRegexFilter(filter, index) {
            const prefix = index !== undefined ? `Invalid RegexFilter at index ${index}` : 'Invalid RegexFilter';
            // Validate name: between 1 and 100 characters
            if (filter.name !== undefined && !aws_cdk_lib_1.Token.isUnresolved(filter.name)) {
                if (filter.name.length < 1) {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: The field name is ${filter.name.length} characters long but must be at least 1 characters`, this);
                }
                if (filter.name.length > 100) {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: The field name is ${filter.name.length} characters long but must be less than or equal to 100 characters`, this);
                }
            }
            // Validate description: between 1 and 1000 characters (if provided)
            if (filter.description !== undefined && !aws_cdk_lib_1.Token.isUnresolved(filter.description)) {
                if (filter.description.length < 1) {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: The field description is ${filter.description.length} characters long but must be at least 1 characters`, this);
                }
                if (filter.description.length > 1000) {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: The field description is ${filter.description.length} characters long but must be less than or equal to 1000 characters`, this);
                }
            }
            // Validate pattern: at least one character
            if (filter.pattern !== undefined && !aws_cdk_lib_1.Token.isUnresolved(filter.pattern)) {
                if (filter.pattern.length < 1) {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: The field pattern is ${filter.pattern.length} characters long but must be at least 1 characters`, this);
                }
            }
            // Validate action: must be a valid GuardrailAction value
            if (filter.action !== undefined && !aws_cdk_lib_1.Token.isUnresolved(filter.action) && !Object.values(filters.GuardrailAction).includes(filter.action)) {
                throw new aws_cdk_lib_1.ValidationError(`${prefix}: action must be a valid GuardrailAction value`, this);
            }
            // Validate inputAction: must be a valid GuardrailAction value (if provided)
            if (filter.inputAction !== undefined &&
                !aws_cdk_lib_1.Token.isUnresolved(filter.inputAction) &&
                !Object.values(filters.GuardrailAction).includes(filter.inputAction)) {
                throw new aws_cdk_lib_1.ValidationError(`${prefix}: inputAction must be a valid GuardrailAction value`, this);
            }
            // Validate outputAction: must be a valid GuardrailAction value (if provided)
            if (filter.outputAction !== undefined &&
                !aws_cdk_lib_1.Token.isUnresolved(filter.outputAction) &&
                !Object.values(filters.GuardrailAction).includes(filter.outputAction)) {
                throw new aws_cdk_lib_1.ValidationError(`${prefix}: outputAction must be a valid GuardrailAction value`, this);
            }
            // Validate inputEnabled: must be a boolean (if provided)
            if (filter.inputEnabled !== undefined &&
                !aws_cdk_lib_1.Token.isUnresolved(filter.inputEnabled) &&
                typeof filter.inputEnabled !== 'boolean') {
                throw new aws_cdk_lib_1.ValidationError(`${prefix}: inputEnabled must be a boolean value`, this);
            }
            // Validate outputEnabled: must be a boolean (if provided)
            if (filter.outputEnabled !== undefined &&
                !aws_cdk_lib_1.Token.isUnresolved(filter.outputEnabled) &&
                typeof filter.outputEnabled !== 'boolean') {
                throw new aws_cdk_lib_1.ValidationError(`${prefix}: outputEnabled must be a boolean value`, this);
            }
            // Apply default values for optional properties if not provided
            if (filter.inputAction === undefined) {
                filter.inputAction = filters.GuardrailAction.BLOCK;
            }
            if (filter.inputEnabled === undefined) {
                filter.inputEnabled = true;
            }
            if (filter.outputAction === undefined) {
                filter.outputAction = filters.GuardrailAction.BLOCK;
            }
            if (filter.outputEnabled === undefined) {
                filter.outputEnabled = true;
            }
        }
        /**
         * Validates a messaging property (blockedInputMessaging or blockedOutputsMessaging).
         * @param value The messaging value to validate.
         * @param propertyName The name of the property being validated.
         */
        validateMessagingProperty(value, propertyName) {
            if (value !== undefined && !aws_cdk_lib_1.Token.isUnresolved(value)) {
                if (value.length < 1) {
                    throw new aws_cdk_lib_1.ValidationError(`Invalid ${propertyName}: The field ${propertyName} is ${value.length} characters long but must be at least 1 characters`, this);
                }
                if (value.length > 500) {
                    throw new aws_cdk_lib_1.ValidationError(`Invalid ${propertyName}: The field ${propertyName} is ${value.length} characters long but must be less than or equal to 500 characters`, this);
                }
            }
        }
        /**
         * Validates that cross-region configuration is provided when STANDARD tier is used.
         * @param props The guardrail properties to validate.
         */
        validateTierConfiguration(props) {
            const contentTierConfig = props.contentFiltersTierConfig ?? filters.TierConfig.CLASSIC;
            const topicsTierConfig = props.topicsTierConfig ?? filters.TierConfig.CLASSIC;
            const hasCrossRegionConfig = props.crossRegionConfig !== undefined;
            // Check if STANDARD tier is used for content filters
            if (contentTierConfig === filters.TierConfig.STANDARD && !hasCrossRegionConfig) {
                throw new aws_cdk_lib_1.ValidationError('Cross-region configuration is required when using STANDARD tier for content filters. ' +
                    'Please provide a crossRegionConfig property with a valid guardrailProfileArn.', this);
            }
            // Check if STANDARD tier is used for topic filters
            if (topicsTierConfig === filters.TierConfig.STANDARD && !hasCrossRegionConfig) {
                throw new aws_cdk_lib_1.ValidationError('Cross-region configuration is required when using STANDARD tier for topic filters. ' +
                    'Please provide a crossRegionConfig property with a valid guardrailProfileArn.', this);
            }
        }
        /**
         * Validates content filters array and applies default values for optional properties.
         * @param contentFilters The content filters to validate.
         */
        validateContentFilters(contentFilters) {
            if (!contentFilters)
                return;
            contentFilters.forEach((filter, index) => {
                const prefix = `Invalid ContentFilter at index ${index}`;
                // Validate that the filter has required properties
                if (!filter.type) {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: type is required`, this);
                }
                // Validate input strength
                if (filter.inputStrength !== undefined && !aws_cdk_lib_1.Token.isUnresolved(filter.inputStrength)) {
                    if (!Object.values(filters.ContentFilterStrength).includes(filter.inputStrength)) {
                        throw new aws_cdk_lib_1.ValidationError(`${prefix}: inputStrength must be a valid ContentFilterStrength value`, this);
                    }
                }
                // Validate output strength
                if (filter.outputStrength !== undefined && !aws_cdk_lib_1.Token.isUnresolved(filter.outputStrength)) {
                    if (!Object.values(filters.ContentFilterStrength).includes(filter.outputStrength)) {
                        throw new aws_cdk_lib_1.ValidationError(`${prefix}: outputStrength must be a valid ContentFilterStrength value`, this);
                    }
                }
                // Validate input modalities
                if (filter.inputModalities) {
                    filter.inputModalities.forEach((modality, modalityIndex) => {
                        if (!Object.values(filters.ModalityType).includes(modality)) {
                            throw new aws_cdk_lib_1.ValidationError(`${prefix}: inputModalities[${modalityIndex}] must be a valid ModalityType value`, this);
                        }
                    });
                }
                // Validate output modalities
                if (filter.outputModalities) {
                    filter.outputModalities.forEach((modality, modalityIndex) => {
                        if (!Object.values(filters.ModalityType).includes(modality)) {
                            throw new aws_cdk_lib_1.ValidationError(`${prefix}: outputModalities[${modalityIndex}] must be a valid ModalityType value`, this);
                        }
                    });
                }
                // Validate inputAction: must be a valid GuardrailAction value (if provided)
                if (filter.inputAction !== undefined &&
                    !aws_cdk_lib_1.Token.isUnresolved(filter.inputAction) &&
                    !Object.values(filters.GuardrailAction).includes(filter.inputAction)) {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: inputAction must be a valid GuardrailAction value`, this);
                }
                // Validate outputAction: must be a valid GuardrailAction value (if provided)
                if (filter.outputAction !== undefined &&
                    !aws_cdk_lib_1.Token.isUnresolved(filter.outputAction) &&
                    !Object.values(filters.GuardrailAction).includes(filter.outputAction)) {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: outputAction must be a valid GuardrailAction value`, this);
                }
                // Validate inputEnabled: must be a boolean (if provided)
                if (filter.inputEnabled !== undefined &&
                    !aws_cdk_lib_1.Token.isUnresolved(filter.inputEnabled) &&
                    typeof filter.inputEnabled !== 'boolean') {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: inputEnabled must be a boolean value`, this);
                }
                // Validate outputEnabled: must be a boolean (if provided)
                if (filter.outputEnabled !== undefined &&
                    !aws_cdk_lib_1.Token.isUnresolved(filter.outputEnabled) &&
                    typeof filter.outputEnabled !== 'boolean') {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: outputEnabled must be a boolean value`, this);
                }
                // Apply default values for optional properties if not provided
                if (filter.inputAction === undefined) {
                    filter.inputAction = filters.GuardrailAction.BLOCK;
                }
                if (filter.inputEnabled === undefined) {
                    filter.inputEnabled = true;
                }
                if (filter.outputAction === undefined) {
                    filter.outputAction = filters.GuardrailAction.BLOCK;
                }
                if (filter.outputEnabled === undefined) {
                    filter.outputEnabled = true;
                }
            });
        }
        /**
         * Validates PII filters array and applies default values for optional properties.
         * @param piiFilters The PII filters to validate.
         */
        validatePiiFilters(piiFilters) {
            if (!piiFilters)
                return;
            piiFilters.forEach((filter, index) => {
                const prefix = `Invalid PIIFilter at index ${index}`;
                // Validate that the filter has required properties
                if (!filter.type) {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: type is required`, this);
                }
                if (!filter.action) {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: action is required`, this);
                }
                // Validate action values
                if (!aws_cdk_lib_1.Token.isUnresolved(filter.action) && !Object.values(filters.GuardrailAction).includes(filter.action)) {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: action must be a valid GuardrailAction value`, this);
                }
                if (filter.inputAction &&
                    !aws_cdk_lib_1.Token.isUnresolved(filter.inputAction) &&
                    !Object.values(filters.GuardrailAction).includes(filter.inputAction)) {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: inputAction must be a valid GuardrailAction value`, this);
                }
                if (filter.outputAction &&
                    !aws_cdk_lib_1.Token.isUnresolved(filter.outputAction) &&
                    !Object.values(filters.GuardrailAction).includes(filter.outputAction)) {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: outputAction must be a valid GuardrailAction value`, this);
                }
                // Validate inputEnabled: must be a boolean (if provided)
                if (filter.inputEnabled !== undefined &&
                    !aws_cdk_lib_1.Token.isUnresolved(filter.inputEnabled) &&
                    typeof filter.inputEnabled !== 'boolean') {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: inputEnabled must be a boolean value`, this);
                }
                // Validate outputEnabled: must be a boolean (if provided)
                if (filter.outputEnabled !== undefined &&
                    !aws_cdk_lib_1.Token.isUnresolved(filter.outputEnabled) &&
                    typeof filter.outputEnabled !== 'boolean') {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: outputEnabled must be a boolean value`, this);
                }
                // Apply default values for optional properties if not provided
                if (filter.inputAction === undefined) {
                    filter.inputAction = filters.GuardrailAction.BLOCK;
                }
                if (filter.inputEnabled === undefined) {
                    filter.inputEnabled = true;
                }
                if (filter.outputAction === undefined) {
                    filter.outputAction = filters.GuardrailAction.BLOCK;
                }
                if (filter.outputEnabled === undefined) {
                    filter.outputEnabled = true;
                }
            });
        }
        /**
         * Validates regex filters array.
         * @param regexFilters The regex filters to validate.
         */
        validateRegexFilters(regexFilters) {
            if (!regexFilters)
                return;
            regexFilters.forEach((filter, index) => {
                this.validateRegexFilter(filter, index);
            });
        }
        /**
         * Validates denied topics array and applies default values for optional properties.
         * @param deniedTopics The denied topics to validate.
         */
        validateDeniedTopics(deniedTopics) {
            if (!deniedTopics)
                return;
            deniedTopics.forEach((topic, index) => {
                const prefix = `Invalid Topic at index ${index}`;
                // Validate that the topic has required properties
                if (!topic.name) {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: name is required`, this);
                }
                if (!topic.definition) {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: definition is required`, this);
                }
                // Validate name length
                if (!aws_cdk_lib_1.Token.isUnresolved(topic.name) && topic.name.length > 100) {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: name must be 100 characters or less`, this);
                }
                // Validate definition length
                if (!aws_cdk_lib_1.Token.isUnresolved(topic.definition) && topic.definition.length > 1000) {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: definition must be 1000 characters or less`, this);
                }
                // Validate examples if provided
                if (topic.examples) {
                    if (topic.examples.length > 100) {
                        throw new aws_cdk_lib_1.ValidationError(`${prefix}: examples array cannot contain more than 100 examples`, this);
                    }
                    topic.examples.forEach((example, exampleIndex) => {
                        if (!aws_cdk_lib_1.Token.isUnresolved(example) && example.length > 100) {
                            throw new aws_cdk_lib_1.ValidationError(`${prefix}: examples[${exampleIndex}] must be 100 characters or less`, this);
                        }
                    });
                }
                // Validate inputAction: must be a valid GuardrailAction value (if provided)
                if (topic.inputAction !== undefined &&
                    !aws_cdk_lib_1.Token.isUnresolved(topic.inputAction) &&
                    !Object.values(filters.GuardrailAction).includes(topic.inputAction)) {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: inputAction must be a valid GuardrailAction value`, this);
                }
                // Validate outputAction: must be a valid GuardrailAction value (if provided)
                if (topic.outputAction !== undefined &&
                    !aws_cdk_lib_1.Token.isUnresolved(topic.outputAction) &&
                    !Object.values(filters.GuardrailAction).includes(topic.outputAction)) {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: outputAction must be a valid GuardrailAction value`, this);
                }
                // Validate inputEnabled: must be a boolean (if provided)
                if (topic.inputEnabled !== undefined &&
                    !aws_cdk_lib_1.Token.isUnresolved(topic.inputEnabled) &&
                    typeof topic.inputEnabled !== 'boolean') {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: inputEnabled must be a boolean value`, this);
                }
                // Validate outputEnabled: must be a boolean (if provided)
                if (topic.outputEnabled !== undefined &&
                    !aws_cdk_lib_1.Token.isUnresolved(topic.outputEnabled) &&
                    typeof topic.outputEnabled !== 'boolean') {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: outputEnabled must be a boolean value`, this);
                }
                // Apply default values for optional properties if not provided
                if (topic.inputAction === undefined) {
                    topic.inputAction = filters.GuardrailAction.BLOCK;
                }
                if (topic.inputEnabled === undefined) {
                    topic.inputEnabled = true;
                }
                if (topic.outputAction === undefined) {
                    topic.outputAction = filters.GuardrailAction.BLOCK;
                }
                if (topic.outputEnabled === undefined) {
                    topic.outputEnabled = true;
                }
            });
        }
        /**
         * Validates contextual grounding filters array and applies default values for optional properties.
         * @param contextualGroundingFilters The contextual grounding filters to validate.
         */
        validateContextualGroundingFilters(contextualGroundingFilters) {
            if (!contextualGroundingFilters)
                return;
            contextualGroundingFilters.forEach((filter, index) => {
                const prefix = `Invalid ContextualGroundingFilter at index ${index}`;
                // Validate that the filter has required properties
                if (!filter.type) {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: type is required`, this);
                }
                if (filter.threshold === undefined) {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: threshold is required`, this);
                }
                // Validate type
                if (!Object.values(filters.ContextualGroundingFilterType).includes(filter.type)) {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: type must be a valid ContextualGroundingFilterType value`, this);
                }
                // Validate threshold range
                if (!aws_cdk_lib_1.Token.isUnresolved(filter.threshold)) {
                    if (filter.threshold < 0 || filter.threshold > 0.99) {
                        throw new aws_cdk_lib_1.ValidationError(`${prefix}: threshold must be between 0 and 0.99`, this);
                    }
                }
                // Validate action: must be a valid GuardrailAction value (if provided)
                if (filter.action !== undefined &&
                    !aws_cdk_lib_1.Token.isUnresolved(filter.action) &&
                    !Object.values(filters.GuardrailAction).includes(filter.action)) {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: action must be a valid GuardrailAction value`, this);
                }
                // Validate enabled: must be a boolean (if provided)
                if (filter.enabled !== undefined &&
                    !aws_cdk_lib_1.Token.isUnresolved(filter.enabled) &&
                    typeof filter.enabled !== 'boolean') {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: enabled must be a boolean value`, this);
                }
                // Apply default values for optional properties if not provided
                if (filter.action === undefined) {
                    filter.action = filters.GuardrailAction.BLOCK;
                }
                if (filter.enabled === undefined) {
                    filter.enabled = true;
                }
            });
        }
        /**
         * Validates word filters array and applies default values for optional properties.
         * @param wordFilters The word filters to validate.
         */
        validateWordFilters(wordFilters) {
            if (!wordFilters)
                return;
            wordFilters.forEach((filter, index) => {
                const prefix = `Invalid WordFilter at index ${index}`;
                // Validate that the filter has required properties
                if (!filter.text) {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: text is required`, this);
                }
                // Validate text length
                if (!aws_cdk_lib_1.Token.isUnresolved(filter.text) && filter.text.length > 100) {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: text must be 100 characters or less`, this);
                }
                // Validate inputAction: must be a valid GuardrailAction value (if provided)
                if (filter.inputAction !== undefined &&
                    !aws_cdk_lib_1.Token.isUnresolved(filter.inputAction) &&
                    !Object.values(filters.GuardrailAction).includes(filter.inputAction)) {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: inputAction must be a valid GuardrailAction value`, this);
                }
                // Validate outputAction: must be a valid GuardrailAction value (if provided)
                if (filter.outputAction !== undefined &&
                    !aws_cdk_lib_1.Token.isUnresolved(filter.outputAction) &&
                    !Object.values(filters.GuardrailAction).includes(filter.outputAction)) {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: outputAction must be a valid GuardrailAction value`, this);
                }
                // Validate inputEnabled: must be a boolean (if provided)
                if (filter.inputEnabled !== undefined &&
                    !aws_cdk_lib_1.Token.isUnresolved(filter.inputEnabled) &&
                    typeof filter.inputEnabled !== 'boolean') {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: inputEnabled must be a boolean value`, this);
                }
                // Validate outputEnabled: must be a boolean (if provided)
                if (filter.outputEnabled !== undefined &&
                    !aws_cdk_lib_1.Token.isUnresolved(filter.outputEnabled) &&
                    typeof filter.outputEnabled !== 'boolean') {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: outputEnabled must be a boolean value`, this);
                }
                // Apply default values for optional properties if not provided
                if (filter.inputAction === undefined) {
                    filter.inputAction = filters.GuardrailAction.BLOCK;
                }
                if (filter.inputEnabled === undefined) {
                    filter.inputEnabled = true;
                }
                if (filter.outputAction === undefined) {
                    filter.outputAction = filters.GuardrailAction.BLOCK;
                }
                if (filter.outputEnabled === undefined) {
                    filter.outputEnabled = true;
                }
            });
        }
        /**
         * Validates managed word list filters array and applies default values for optional properties.
         * @param managedWordListFilters The managed word list filters to validate.
         */
        validateManagedWordListFilters(managedWordListFilters) {
            if (!managedWordListFilters)
                return;
            managedWordListFilters.forEach((filter, index) => {
                const prefix = `Invalid ManagedWordFilter at index ${index}`;
                // Validate type: must be a valid ManagedWordFilterType value (if provided)
                if (filter.type !== undefined && !Object.values(filters.ManagedWordFilterType).includes(filter.type)) {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: type must be a valid ManagedWordFilterType value`, this);
                }
                // Validate inputAction: must be a valid GuardrailAction value (if provided)
                if (filter.inputAction !== undefined &&
                    !aws_cdk_lib_1.Token.isUnresolved(filter.inputAction) &&
                    !Object.values(filters.GuardrailAction).includes(filter.inputAction)) {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: inputAction must be a valid GuardrailAction value`, this);
                }
                // Validate outputAction: must be a valid GuardrailAction value (if provided)
                if (filter.outputAction !== undefined &&
                    !aws_cdk_lib_1.Token.isUnresolved(filter.outputAction) &&
                    !Object.values(filters.GuardrailAction).includes(filter.outputAction)) {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: outputAction must be a valid GuardrailAction value`, this);
                }
                // Validate inputEnabled: must be a boolean (if provided)
                if (filter.inputEnabled !== undefined &&
                    !aws_cdk_lib_1.Token.isUnresolved(filter.inputEnabled) &&
                    typeof filter.inputEnabled !== 'boolean') {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: inputEnabled must be a boolean value`, this);
                }
                // Validate outputEnabled: must be a boolean (if provided)
                if (filter.outputEnabled !== undefined &&
                    !aws_cdk_lib_1.Token.isUnresolved(filter.outputEnabled) &&
                    typeof filter.outputEnabled !== 'boolean') {
                    throw new aws_cdk_lib_1.ValidationError(`${prefix}: outputEnabled must be a boolean value`, this);
                }
                // Apply default values for optional properties if not provided
                if (filter.type === undefined) {
                    filter.type = filters.ManagedWordFilterType.PROFANITY;
                }
                if (filter.inputAction === undefined) {
                    filter.inputAction = filters.GuardrailAction.BLOCK;
                }
                if (filter.inputEnabled === undefined) {
                    filter.inputEnabled = true;
                }
                if (filter.outputAction === undefined) {
                    filter.outputAction = filters.GuardrailAction.BLOCK;
                }
                if (filter.outputEnabled === undefined) {
                    filter.outputEnabled = true;
                }
            });
        }
        /**
         * Validates a single content filter and applies default values for optional properties.
         * @param filter The content filter to validate.
         */
        validateSingleContentFilter(filter) {
            // Use the existing validation logic but catch and re-throw with a clearer error message
            try {
                this.validateContentFilters([filter]);
            }
            catch (error) {
                if (error instanceof aws_cdk_lib_1.ValidationError) {
                    // Replace "at index 0" with a clearer message for single filter validation
                    const message = error.message.replace(' at index 0', '');
                    throw new aws_cdk_lib_1.ValidationError(message, this);
                }
                throw error;
            }
        }
        /**
         * Validates a single PII filter and applies default values for optional properties.
         * @param filter The PII filter to validate.
         */
        validateSinglePiiFilter(filter) {
            // Use the existing validation logic but catch and re-throw with a clearer error message
            try {
                this.validatePiiFilters([filter]);
            }
            catch (error) {
                if (error instanceof aws_cdk_lib_1.ValidationError) {
                    // Replace "at index 0" with a clearer message for single filter validation
                    const message = error.message.replace(' at index 0', '');
                    throw new aws_cdk_lib_1.ValidationError(message, this);
                }
                throw error;
            }
        }
        /**
         * Validates a single regex filter and applies default values for optional properties.
         * @param filter The regex filter to validate.
         */
        validateSingleRegexFilter(filter) {
            // Use the existing validation logic but catch and re-throw with a clearer error message
            try {
                this.validateRegexFilters([filter]);
            }
            catch (error) {
                if (error instanceof aws_cdk_lib_1.ValidationError) {
                    // Replace "at index 0" with a clearer message for single filter validation
                    const message = error.message.replace(' at index 0', '');
                    throw new aws_cdk_lib_1.ValidationError(message, this);
                }
                throw error;
            }
        }
        /**
         * Validates a single denied topic and applies default values for optional properties.
         * @param filter The denied topic to validate.
         */
        validateSingleDeniedTopic(filter) {
            // Use the existing validation logic but catch and re-throw with a clearer error message
            try {
                this.validateDeniedTopics([filter]);
            }
            catch (error) {
                if (error instanceof aws_cdk_lib_1.ValidationError) {
                    // Replace "at index 0" with a clearer message for single filter validation
                    const message = error.message.replace(' at index 0', '');
                    throw new aws_cdk_lib_1.ValidationError(message, this);
                }
                throw error;
            }
        }
        /**
         * Validates a single contextual grounding filter and applies default values for optional properties.
         * @param filter The contextual grounding filter to validate.
         */
        validateSingleContextualGroundingFilter(filter) {
            // Use the existing validation logic but catch and re-throw with a clearer error message
            try {
                this.validateContextualGroundingFilters([filter]);
            }
            catch (error) {
                if (error instanceof aws_cdk_lib_1.ValidationError) {
                    // Replace "at index 0" with a clearer message for single filter validation
                    const message = error.message.replace(' at index 0', '');
                    throw new aws_cdk_lib_1.ValidationError(message, this);
                }
                throw error;
            }
        }
        /**
         * Validates a single word filter and applies default values for optional properties.
         * @param filter The word filter to validate.
         */
        validateSingleWordFilter(filter) {
            // Use the existing validation logic but catch and re-throw with a clearer error message
            try {
                this.validateWordFilters([filter]);
            }
            catch (error) {
                if (error instanceof aws_cdk_lib_1.ValidationError) {
                    // Replace "at index 0" with a clearer message for single filter validation
                    const message = error.message.replace(' at index 0', '');
                    throw new aws_cdk_lib_1.ValidationError(message, this);
                }
                throw error;
            }
        }
        /**
         * Validates a single managed word list filter and applies default values for optional properties.
         * @param filter The managed word list filter to validate.
         */
        validateSingleManagedWordListFilter(filter) {
            // Use the existing validation logic but catch and re-throw with a clearer error message
            try {
                this.validateManagedWordListFilters([filter]);
            }
            catch (error) {
                if (error instanceof aws_cdk_lib_1.ValidationError) {
                    // Replace "at index 0" with a clearer message for single filter validation
                    const message = error.message.replace(' at index 0', '');
                    throw new aws_cdk_lib_1.ValidationError(message, this);
                }
                throw error;
            }
        }
        static {
            __runInitializers(_classThis, _classExtraInitializers);
        }
    };
    return Guardrail = _classThis;
})();
exports.Guardrail = Guardrail;
