// @aws-cdk/aws-bedrock-alpha/bedrock/guardrails/guardrail-filters.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContextualGroundingFilterType = exports.UKSpecificPIIType = exports.CanadaSpecificPIIType = exports.USASpecificPIIType = exports.InformationTechnologyPIIType = exports.FinancePIIType = exports.GeneralPIIType = exports.PIIType = exports.ManagedWordFilterType = exports.Topic = exports.ContentFilterType = exports.ModalityType = exports.ContentFilterStrength = exports.TierConfig = exports.GuardrailAction = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
/**
 * Guardrail action when a sensitive entity is detected.
 */
var GuardrailAction;
(function (GuardrailAction) {
    /**
     * If sensitive information is detected in the prompt or response, the guardrail
     * blocks all the content and returns a message that you configure.
     */
    GuardrailAction["BLOCK"] = "BLOCK";
    /**
     * If sensitive information is detected in the model response, the guardrail masks
     * it with an identifier, the sensitive information is masked and replaced with
     * identifier tags (for example: [NAME-1], [NAME-2], [EMAIL-1], etc.).
     */
    GuardrailAction["ANONYMIZE"] = "ANONYMIZE";
    /**
     * Do not take any action.
     */
    GuardrailAction["NONE"] = "NONE";
})(GuardrailAction || (exports.GuardrailAction = GuardrailAction = {}));
/******************************************************************************
 *                             TIER CONFIG
*****************************************************************************/
var TierConfig;
(function (TierConfig) {
    /**
     * Provides established guardrails functionality supporting English, French, and Spanish languages.
     */
    TierConfig["CLASSIC"] = "CLASSIC";
    /**
     * Provides a more robust solution than the CLASSIC tier and has more comprehensive language support. This tier requires that your guardrail use cross-Region inference.
     */
    TierConfig["STANDARD"] = "STANDARD";
})(TierConfig || (exports.TierConfig = TierConfig = {}));
/******************************************************************************
 *                             CONTENT FILTERS
 *****************************************************************************/
/**
 * The strength of the content filter. As you increase the filter strength,
 * the likelihood of filtering harmful content increases and the probability
 * of seeing harmful content in your application reduces.
 */
var ContentFilterStrength;
(function (ContentFilterStrength) {
    /**
     * No content filtering applied.
     */
    ContentFilterStrength["NONE"] = "NONE";
    /**
     * Low strength content filtering - minimal filtering of harmful content.
     */
    ContentFilterStrength["LOW"] = "LOW";
    /**
     * Medium strength content filtering - balanced filtering of harmful content.
     */
    ContentFilterStrength["MEDIUM"] = "MEDIUM";
    /**
     * High strength content filtering - aggressive filtering of harmful content.
     */
    ContentFilterStrength["HIGH"] = "HIGH";
})(ContentFilterStrength || (exports.ContentFilterStrength = ContentFilterStrength = {}));
/**
 * The type of modality that can be used in content filters.
 */
var ModalityType;
(function (ModalityType) {
    /**
     * Text modality for content filters.
     */
    ModalityType["TEXT"] = "TEXT";
    /**
     * Image modality for content filters.
     */
    ModalityType["IMAGE"] = "IMAGE";
})(ModalityType || (exports.ModalityType = ModalityType = {}));
/**
 * The type of harmful category usable in a content filter.
 */
var ContentFilterType;
(function (ContentFilterType) {
    /**
     * Describes input prompts and model responses that indicates sexual interest, activity,
     * or arousal using direct or indirect references to body parts, physical traits, or sex.
     */
    ContentFilterType["SEXUAL"] = "SEXUAL";
    /**
     * Describes input prompts and model responses that includes glorification of or threats
     * to inflict physical pain, hurt, or injury toward a person, group or thing.
     */
    ContentFilterType["VIOLENCE"] = "VIOLENCE";
    /**
     * Describes input prompts and model responses that discriminate, criticize, insult,
     * denounce, or dehumanize a person or group on the basis of an identity (such as race,
     * ethnicity, gender, religion, sexual orientation, ability, and national origin).
     */
    ContentFilterType["HATE"] = "HATE";
    /**
     * Describes input prompts and model responses that includes demeaning, humiliating,
     * mocking, insulting, or belittling language. This type of language is also labeled
     * as bullying.
     */
    ContentFilterType["INSULTS"] = "INSULTS";
    /**
     * Describes input prompts and model responses that seeks or provides information
     * about engaging in misconduct activity, or harming, defrauding, or taking advantage
     * of a person, group or institution.
     */
    ContentFilterType["MISCONDUCT"] = "MISCONDUCT";
    /**
     * Enable to detect and block user inputs attempting to override system instructions.
     * To avoid misclassifying system prompts as a prompt attack and ensure that the filters
     * are selectively applied to user inputs, use input tagging.
     */
    ContentFilterType["PROMPT_ATTACK"] = "PROMPT_ATTACK";
})(ContentFilterType || (exports.ContentFilterType = ContentFilterType = {}));
/**
 * Represents predefined topics that can be used to filter content.
 */
class Topic {
    static [JSII_RTTI_SYMBOL_1] = { fqn: "@aws-cdk/aws-bedrock-alpha.Topic", version: "2.235.0-alpha.0" };
    /**
     * Filter for financial advice and investment recommendations.
     */
    static FINANCIAL_ADVICE = new Topic({
        name: 'Financial_Advice',
        definition: "'Discussions that involve providing guidance, recommendations, or suggestions related to managing, investing, or handling finances, investments, or assets.",
        examples: [
            'Can you suggest some good stocks to invest in right now?',
            "What's the best way to save for retirement?",
            'Should I put my money in a high-risk investment?',
            'How can I maximize my returns on investments?',
            'Is it a good time to buy real estate?',
        ],
        inputAction: GuardrailAction.BLOCK,
        inputEnabled: true,
        outputAction: GuardrailAction.BLOCK,
        outputEnabled: true,
    });
    /**
     * Filter for inappropriate or explicit content.
     */
    static INAPPROPRIATE_CONTENT = new Topic({
        name: 'Inappropriate_Content',
        definition: 'Conversations that express views, opinions, or endorsements related to political parties, political ideologies, elections, or political figures.',
        examples: [
            "What's your stance on the current government policies?",
            'Do you support party X or Y in the upcoming election?',
            'Can you tell me which political party is better?',
            'What do you think about the new policy introduced by the president?',
            'Should I vote for this candidate?',
        ],
        inputAction: GuardrailAction.BLOCK,
        inputEnabled: true,
        outputAction: GuardrailAction.BLOCK,
        outputEnabled: true,
    });
    /**
     * Filter for legal advice and recommendations.
     */
    static LEGAL_ADVICE = new Topic({
        name: 'Legal_Advice',
        definition: 'Providing recommendations, diagnosis, treatment options, or guidance on medical conditions, symptoms, medications, or health-related issues.',
        examples: [
            'What should I do to treat a persistent cough?',
            'Can you recommend some medication for my headache?',
            'What are the symptoms of a specific medical condition?',
            'Is this drug effective for treating my illness?',
            "Do I need to see a doctor for this pain I'm experiencing?",
        ],
        inputAction: GuardrailAction.BLOCK,
        inputEnabled: true,
        outputAction: GuardrailAction.BLOCK,
        outputEnabled: true,
    });
    /**
     * Filter for medical advice and health recommendations.
     */
    static MEDICAL_ADVICE = new Topic({
        name: 'Medical_Advice',
        definition: 'Any discussions or references that include hate speech, discriminatory remarks, sexual content, or explicit language.',
        examples: [
            'Why are people from X country so [discriminatory remark]?',
            'Can you tell me a dirty joke?',
            '[Use of explicit language]',
            'This service is as bad as [hate speech].',
            'Do you have any adult content or products?',
        ],
        inputAction: GuardrailAction.BLOCK,
        inputEnabled: true,
        outputAction: GuardrailAction.BLOCK,
        outputEnabled: true,
    });
    /**
     * Filter for political advice and recommendations.
     */
    static POLITICAL_ADVICE = new Topic({
        name: 'Political_Advice',
        definition: 'Offering guidance or suggestions on legal matters, legal actions, interpretation of laws, or legal rights and responsibilities.',
        examples: [
            'Can I sue someone for this?',
            'What are my legal rights in this situation?',
            'Is this action against the law?',
            'What should I do to file a legal complaint?',
            'Can you explain this law to me?',
        ],
        inputAction: GuardrailAction.BLOCK,
        inputEnabled: true,
        outputAction: GuardrailAction.BLOCK,
        outputEnabled: true,
    });
    /**
     * Create a custom topic filter.
     * @param props Properties for the custom topic filter
     */
    static custom(props) {
        try {
            jsiiDeprecationWarnings._aws_cdk_aws_bedrock_alpha_CustomTopicProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.custom);
            }
            throw error;
        }
        return new Topic(props);
    }
    /**
     * The name of the topic to deny.
     */
    name;
    /**
     * Definition of the topic.
     */
    definition;
    /**
     * Representative phrases that refer to the topic.
     */
    examples;
    /**
     * The action to take when a topic is detected in the input.
     * @default GuardrailAction.BLOCK
     */
    inputAction;
    /**
     * Whether the topic filter is enabled for input.
     * @default true
     */
    inputEnabled;
    /**
     * The action to take when a topic is detected in the output.
     * @default GuardrailAction.BLOCK
     */
    outputAction;
    /**
     * Whether the topic filter is enabled for output.
     * @default true
     */
    outputEnabled;
    constructor(props) {
        try {
            jsiiDeprecationWarnings._aws_cdk_aws_bedrock_alpha_CustomTopicProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, Topic);
            }
            throw error;
        }
        // Validate examples field constraints
        const examplesValidationRules = [
            {
                condition: (p) => !p.examples || p.examples.length < 1,
                message: () => 'examples field must contain at least 1 example',
            },
            {
                condition: (p) => p.examples && p.examples.length > 100,
                message: () => 'examples field cannot contain more than 100 examples',
            },
        ];
        // Note: We can't use validateAllProps here since we don't have a Construct scope
        // Instead, we'll validate directly
        for (const rule of examplesValidationRules) {
            if (rule.condition(props)) {
                throw new Error(rule.message(props));
            }
        }
        this.name = props.name;
        this.definition = props.definition;
        this.examples = props.examples;
        this.inputAction = props.inputAction;
        this.inputEnabled = props.inputEnabled;
        this.outputAction = props.outputAction;
        this.outputEnabled = props.outputEnabled;
    }
}
exports.Topic = Topic;
/**
 * Managed word list filter types supported by Amazon Bedrock.
 */
var ManagedWordFilterType;
(function (ManagedWordFilterType) {
    /**
     * Filter for profanity and explicit language.
     */
    ManagedWordFilterType["PROFANITY"] = "PROFANITY";
})(ManagedWordFilterType || (exports.ManagedWordFilterType = ManagedWordFilterType = {}));
/******************************************************************************
   *                   SENSITIVE INFORMATION FILTERS - PII
   *****************************************************************************/
/**
 * Abstract base class for all PII types.
 */
class PIIType {
    static [JSII_RTTI_SYMBOL_1] = { fqn: "@aws-cdk/aws-bedrock-alpha.PIIType", version: "2.235.0-alpha.0" };
    /**
     * The string value of the PII type.
     */
    value;
    /**
     * The string value of the PII type.
     */
    constructor(value) {
        this.value = value;
    }
    /**
     * Returns the string representation of the PII type.
     * @returns The string value of the PII type.
     */
    toString() { return this.value; }
}
exports.PIIType = PIIType;
/**
 * Types of PII that are general, and not domain-specific.
 */
class GeneralPIIType extends PIIType {
    static [JSII_RTTI_SYMBOL_1] = { fqn: "@aws-cdk/aws-bedrock-alpha.GeneralPIIType", version: "2.235.0-alpha.0" };
    /**
     * A physical address, such as "100 Main Street, Anytown, USA" or "Suite #12,
     * Building 123". An address can include information such as the street, building,
     * location, city, state, country, county, zip code, precinct, and neighborhood.
     */
    static ADDRESS = new GeneralPIIType('ADDRESS');
    /**
     * An individual's age, including the quantity and unit of time.
     */
    static AGE = new GeneralPIIType('AGE');
    /**
     * The number assigned to a driver's license, which is an official document
     * permitting an individual to operate one or more motorized vehicles on a
     * public road. A driver's license number consists of alphanumeric characters.
     */
    static DRIVER_ID = new GeneralPIIType('DRIVER_ID');
    /**
     * An email address, such as marymajor@email.com.
     */
    static EMAIL = new GeneralPIIType('EMAIL');
    /**
     * A license plate for a vehicle is issued by the state or country where the
     * vehicle is registered. The format for passenger vehicles is typically five
     * to eight digits, consisting of upper-case letters and numbers. The format
     * varies depending on the location of the issuing state or country.
     */
    static LICENSE_PLATE = new GeneralPIIType('LICENSE_PLATE');
    /**
     * An individual's name. This entity type does not include titles, such as Dr.,
     *  Mr., Mrs., or Miss.
     */
    static NAME = new GeneralPIIType('NAME');
    /**
     * An alphanumeric string that is used as a password, such as "*very20special#pass*".
     */
    static PASSWORD = new GeneralPIIType('PASSWORD');
    /**
     * A phone number. This entity type also includes fax and pager numbers.
     */
    static PHONE = new GeneralPIIType('PHONE');
    /**
     * A user name that identifies an account, such as a login name, screen name,
     * nick name, or handle.
     */
    static USERNAME = new GeneralPIIType('USERNAME');
    /**
     * A Vehicle Identification Number (VIN) uniquely identifies a vehicle. VIN
     * content and format are defined in the ISO 3779 specification. Each country
     * has specific codes and formats for VINs.
     */
    static VEHICLE_IDENTIFICATION_NUMBER = new GeneralPIIType('VEHICLE_IDENTIFICATION_NUMBER');
    constructor(value) { super(value); }
}
exports.GeneralPIIType = GeneralPIIType;
/**
 * Types of PII in the domain of Finance.
 */
class FinancePIIType extends PIIType {
    static [JSII_RTTI_SYMBOL_1] = { fqn: "@aws-cdk/aws-bedrock-alpha.FinancePIIType", version: "2.235.0-alpha.0" };
    /**
     * A three-digit card verification code (CVV) that is present on VISA, MasterCard,
     * and Discover credit and debit cards. For American Express credit or debit cards,
     * the CVV is a four-digit numeric code.
     */
    static CREDIT_DEBIT_CARD_CVV = new FinancePIIType('CREDIT_DEBIT_CARD_CVV');
    /**
     * The expiration date for a credit or debit card. This number is usually four digits
     * long and is often formatted as month/year or MM/YY. Guardrails recognizes expiration
     * dates such as 01/21, 01/2021, and Jan 2021.
     */
    static CREDIT_DEBIT_CARD_EXPIRY = new FinancePIIType('CREDIT_DEBIT_CARD_EXPIRY');
    /**
     * The number for a credit or debit card. These numbers can vary from 13 to 16 digits
     * in length.
     */
    static CREDIT_DEBIT_CARD_NUMBER = new FinancePIIType('CREDIT_DEBIT_CARD_NUMBER');
    /**
     * A four-digit personal identification number (PIN) with which you can access your
     * bank account.
     */
    static PIN = new FinancePIIType('PIN');
    /**
     * A SWIFT code is a standard format of Bank Identifier Code (BIC) used to specify a
     * particular bank or branch. Banks use these codes for money transfers such as
     * international wire transfers. SWIFT codes consist of eight or 11 characters.
     */
    static SWIFT_CODE = new FinancePIIType('SWIFT_CODE');
    /**
     * An International Bank Account Number (IBAN). It has specific formats in each country.
     */
    static INTERNATIONAL_BANK_ACCOUNT_NUMBER = new FinancePIIType('INTERNATIONAL_BANK_ACCOUNT_NUMBER');
    constructor(value) { super(value); }
}
exports.FinancePIIType = FinancePIIType;
/**
 * Types of PII in the domain of IT (Information Technology).
 */
class InformationTechnologyPIIType extends PIIType {
    static [JSII_RTTI_SYMBOL_1] = { fqn: "@aws-cdk/aws-bedrock-alpha.InformationTechnologyPIIType", version: "2.235.0-alpha.0" };
    /**
     * A web address, such as www.example.com.
     */
    static URL = new InformationTechnologyPIIType('URL');
    /**
     * An IPv4 address, such as 198.51.100.0.
     */
    static IP_ADDRESS = new InformationTechnologyPIIType('IP_ADDRESS');
    /**
     * A media access control (MAC) address assigned to a network interface.
     */
    static MAC_ADDRESS = new InformationTechnologyPIIType('MAC_ADDRESS');
    /**
     * A unique identifier that's associated with a secret access key. You use
     * the access key ID and secret access key to sign programmatic AWS requests
     * cryptographically.
     */
    static AWS_ACCESS_KEY = new InformationTechnologyPIIType('AWS_ACCESS_KEY');
    /**
     * A unique identifier that's associated with a secret access key. You use
     * the access key ID and secret access key to sign programmatic AWS requests
     * cryptographically.
     */
    static AWS_SECRET_KEY = new InformationTechnologyPIIType('AWS_SECRET_KEY');
    constructor(value) { super(value); }
}
exports.InformationTechnologyPIIType = InformationTechnologyPIIType;
/**
 * Types of PII specific to the USA.
 */
class USASpecificPIIType extends PIIType {
    static [JSII_RTTI_SYMBOL_1] = { fqn: "@aws-cdk/aws-bedrock-alpha.USASpecificPIIType", version: "2.235.0-alpha.0" };
    /**
     * A US bank account number, which is typically 10 to 12 digits long.
     */
    static US_BANK_ACCOUNT_NUMBER = new USASpecificPIIType('US_BANK_ACCOUNT_NUMBER');
    /**
     * A US bank account routing number. These are typically nine digits long.
     */
    static US_BANK_ROUTING_NUMBER = new USASpecificPIIType('US_BANK_ROUTING_NUMBER');
    /**
     * A US Individual Taxpayer Identification Number (ITIN) is a nine-digit number
     * that starts with a "9" and contain a "7" or "8" as the fourth digit.
     */
    static US_INDIVIDUAL_TAX_IDENTIFICATION_NUMBER = new USASpecificPIIType('US_INDIVIDUAL_TAX_IDENTIFICATION_NUMBER');
    /**
     * A US passport number. Passport numbers range from six to nine alphanumeric characters.
     */
    static US_PASSPORT_NUMBER = new USASpecificPIIType('US_PASSPORT_NUMBER');
    /**
     * A US Social Security Number (SSN) is a nine-digit number that is issued to US citizens,
     * permanent residents, and temporary working residents.
     */
    static US_SOCIAL_SECURITY_NUMBER = new USASpecificPIIType('US_SOCIAL_SECURITY_NUMBER');
    constructor(value) { super(value); }
}
exports.USASpecificPIIType = USASpecificPIIType;
/**
 * Types of PII specific to Canada.
 */
class CanadaSpecificPIIType extends PIIType {
    static [JSII_RTTI_SYMBOL_1] = { fqn: "@aws-cdk/aws-bedrock-alpha.CanadaSpecificPIIType", version: "2.235.0-alpha.0" };
    /**
     * A Canadian Health Service Number is a 10-digit unique identifier,
     * required for individuals to access healthcare benefits.
     */
    static CA_HEALTH_NUMBER = new CanadaSpecificPIIType('CA_HEALTH_NUMBER');
    /**
     * A Canadian Social Insurance Number (SIN) is a nine-digit unique identifier,
     * required for individuals to access government programs and benefits.
     */
    static CA_SOCIAL_INSURANCE_NUMBER = new CanadaSpecificPIIType('CA_SOCIAL_INSURANCE_NUMBER');
    constructor(value) { super(value); }
}
exports.CanadaSpecificPIIType = CanadaSpecificPIIType;
/**
 * Types of PII specific to the United Kingdom (UK).
 */
class UKSpecificPIIType extends PIIType {
    static [JSII_RTTI_SYMBOL_1] = { fqn: "@aws-cdk/aws-bedrock-alpha.UKSpecificPIIType", version: "2.235.0-alpha.0" };
    /**
     * A UK National Health Service Number is a 10-17 digit number, such as 485 777 3456.
     */
    static UK_NATIONAL_HEALTH_SERVICE_NUMBER = new UKSpecificPIIType('UK_NATIONAL_HEALTH_SERVICE_NUMBER');
    /**
     * A UK National Insurance Number (NINO) provides individuals with access to National
     * Insurance (social security) benefits. It is also used for some purposes in the UK
     * tax system.
     */
    static UK_NATIONAL_INSURANCE_NUMBER = new UKSpecificPIIType('UK_NATIONAL_INSURANCE_NUMBER');
    /**
     * A UK Unique Taxpayer Reference (UTR) is a 10-digit number that identifies a
     * taxpayer or a business.
     */
    static UK_UNIQUE_TAXPAYER_REFERENCE_NUMBER = new UKSpecificPIIType('UK_UNIQUE_TAXPAYER_REFERENCE_NUMBER');
    constructor(value) { super(value); }
}
exports.UKSpecificPIIType = UKSpecificPIIType;
/******************************************************************************
   *                      CONTEXTUAL GROUNDING FILTERS
   *****************************************************************************/
/**
 * The type of contextual grounding filter.
 */
var ContextualGroundingFilterType;
(function (ContextualGroundingFilterType) {
    /**
     * Grounding score represents the confidence that the model response is factually
     * correct and grounded in the source. If the model response has a lower score than
     * the defined threshold, the response will be blocked and the configured blocked
     * message will be returned to the user. A higher threshold level blocks more responses.
     */
    ContextualGroundingFilterType["GROUNDING"] = "GROUNDING";
    /**
     * Relevance score represents the confidence that the model response is relevant
     * to the user's query. If the model response has a lower score than the defined
     * threshold, the response will be blocked and the configured blocked message will
     * be returned to the user. A higher threshold level blocks more responses.
     */
    ContextualGroundingFilterType["RELEVANCE"] = "RELEVANCE";
})(ContextualGroundingFilterType || (exports.ContextualGroundingFilterType = ContextualGroundingFilterType = {}));
