// @cdklabs/generative-ai-cdk-constructs/lib/cdk-lib/aurora-dsql/cluster.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Cluster = exports.ClusterBase = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
const aws_cdk_lib_1 = require("aws-cdk-lib");
const dsql = require("aws-cdk-lib/aws-dsql");
const iam = require("aws-cdk-lib/aws-iam");
// Internal libs
const perms = require("./perms");
const validation_helpers_1 = require("./validation-helpers");
/******************************************************************************
 *                              CONSTANTS
 *****************************************************************************/
/**
 * Minimum length for cluster tags
 * @internal
 */
const CLUSTER_TAG_MIN_LENGTH = 1;
/**
 * Maximum length for browser tags
 * @internal
 */
const CLUSTER_TAG_MAX_LENGTH = 256;
/******************************************************************************
 *                        ABSTRACT BASE CLASS
 *****************************************************************************/
/**
 * Abstract base class for a Aurora DSQL cluster.
 * Contains methods and attributes valid for Aurora DSQL clusters either created with CDK or imported.
 */
class ClusterBase extends aws_cdk_lib_1.Resource {
    constructor(scope, id) {
        super(scope, id);
    }
    /**
     * Grants IAM actions to the IAM Principal
     * @param grantee - The IAM principal to grant permissions to
     * @param actions - The actions to grant
     * @returns An IAM Grant object representing the granted permissions
     */
    grant(grantee, ...actions) {
        return iam.Grant.addToPrincipal({
            grantee: grantee,
            resourceArns: [this.clusterArn],
            actions: actions,
        });
    }
    /**
     * Grants connection authorization to the IAM Principal
     * @see https://docs.aws.amazon.com/aurora-dsql/latest/userguide/authentication-authorization.html#authentication-authorization-iam-policy
     * @param grantee - The IAM principal to grant permissions to
     * @default - Default grant configuration:
     * - actions: ['dsql:DbConnect'] on this.clusterArn
     * @returns An IAM Grant object representing the granted permissions
     */
    grantConnect(grantee) {
        return this.grant(grantee, ...perms.AURORA_DSQL_CONNECT_PERMS);
    }
    /**
     * Grants connection authorization to the IAM Principal
     * @see https://docs.aws.amazon.com/aurora-dsql/latest/userguide/authentication-authorization.html#authentication-authorization-iam-policy
     * @param grantee - The IAM principal to grant permissions to
     * @default - Default grant configuration:
     * - actions: ['dsql:DbConnectAdmin'] on this.clusterArn
     * @returns An IAM Grant object representing the granted permissions
     */
    grantConnectAdmin(grantee) {
        return this.grant(grantee, ...perms.AURORA_DSQL_CONNECT_ADMIN_PERMS);
    }
}
exports.ClusterBase = ClusterBase;
_a = JSII_RTTI_SYMBOL_1;
ClusterBase[_a] = { fqn: "@cdklabs/generative-ai-cdk-constructs.auroraDsql.ClusterBase", version: "0.1.312" };
/******************************************************************************
 *                                Class
 *****************************************************************************/
/**
 * Aurora DSQL cluster resource for AWS Aurora DSQL.
 * You can use this resource to create, modify, and manage clusters.
 * @see https://docs.aws.amazon.com/aurora-dsql/latest/userguide/what-is-aurora-dsql.html
 * @resource AWS::DSQL::Cluster
 */
class Cluster extends ClusterBase {
    /**
     * Static Method for importing an existing Aurora DSQL cluster.
     */
    /**
     * Creates an Aurora DSQL cluster reference from an existing cluster's attributes.
     *
     * @param scope - The construct scope
     * @param id - Identifier of the construct
     * @param attrs - Attributes of the existing cluster
     * @returns An ICluster reference to the existing cluster
     */
    static fromClusterAttributes(scope, id, attrs) {
        class Import extends ClusterBase {
            constructor(s, i) {
                super(s, i);
                this.clusterArn = attrs.clusterArn;
                this.clusterId = aws_cdk_lib_1.Arn.split(attrs.clusterArn, aws_cdk_lib_1.ArnFormat.SLASH_RESOURCE_NAME).resourceName;
                this.creationTime = attrs.creationTime;
                this.status = attrs.status;
                this.vpcEndpointServiceName = attrs.vpcEndpointServiceName;
                this.encryptionKey = attrs.encryptionKey;
            }
        }
        // Return new Cluster
        return new Import(scope, id);
    }
    // ------------------------------------------------------
    // CONSTRUCTOR
    // ------------------------------------------------------
    constructor(scope, id, props) {
        super(scope, id);
        // ------------------------------------------------------
        // Validators
        // ------------------------------------------------------
        /**
         * Validates the multi-region properties
         * @param props The multi-region properties to validate
         * @throws Error if the multi-region properties are invalid
         */
        this._validateMultiRegionProperties = (props) => {
            if (!props) {
                return;
            }
            // The witness Region specified in multiRegionProperties.witnessRegion cannot be the same as the cluster's Region.
            if (props.witnessRegion && !aws_cdk_lib_1.Token.isUnresolved(this.stack.region)) {
                if (props.witnessRegion === this.stack.region) {
                    throw new Error('Witness region cannot be the same as the cluster\'s Region');
                }
            }
            // For the provided peered clusters, they must be in different Regions.
            if (props.clusters && !aws_cdk_lib_1.Token.isUnresolved(this.stack.region)) {
                for (const cluster of props.clusters) {
                    const region = aws_cdk_lib_1.Arn.split(cluster.clusterArn, aws_cdk_lib_1.ArnFormat.SLASH_RESOURCE_NAME).region;
                    if (region === this.stack.region) {
                        throw new Error(`Peered cluster ${cluster.clusterArn} cannot be in the same Region ${this.stack.region} as the cluster's Region`);
                    }
                }
            }
        };
        /**
         * Validates the removal policy
         * @param removalPolicy The removal policy to validate
         * @throws Error if the removal policy is invalid
         */
        this._validateRemovalPolicy = (removalPolicy) => {
            if (removalPolicy && removalPolicy !== aws_cdk_lib_1.RemovalPolicy.DESTROY && removalPolicy !== aws_cdk_lib_1.RemovalPolicy.RETAIN) {
                throw new Error('Invalid removal policy. Only RemovalPolicy.DESTROY and RemovalPolicy.RETAIN are allowed.');
            }
        };
        /**
         * Validates the cluster tags format
         * @param tags The tags object to validate
         * @returns Array of validation error messages, empty if valid
         */
        this._validateClusterTags = (tags) => {
            let errors = [];
            if (!tags) {
                return errors; // Tags are optional
            }
            // Validate each tag key and value
            for (const [key, value] of Object.entries(tags)) {
                errors.push(...(0, validation_helpers_1.validateStringFieldLength)({
                    value: key,
                    fieldName: 'Tag key',
                    minLength: CLUSTER_TAG_MIN_LENGTH,
                    maxLength: CLUSTER_TAG_MAX_LENGTH,
                }));
                // Validate tag key pattern: ^[a-zA-Z0-9\s._:/=+@-]*$
                const validKeyPattern = /^[a-zA-Z0-9\s._:/=+@-]*$/;
                errors.push(...(0, validation_helpers_1.validateFieldPattern)(key, 'Tag key', validKeyPattern));
                // Validate tag value
                errors.push(...(0, validation_helpers_1.validateStringFieldLength)({
                    value: value,
                    fieldName: 'Tag value',
                    minLength: CLUSTER_TAG_MIN_LENGTH,
                    maxLength: CLUSTER_TAG_MAX_LENGTH,
                }));
                // Validate tag value pattern: ^[a-zA-Z0-9\s._:/=+@-]*$
                const validValuePattern = /^[a-zA-Z0-9\s._:/=+@-]*$/;
                errors.push(...(0, validation_helpers_1.validateFieldPattern)(value, 'Tag value', validValuePattern));
            }
            return errors;
        };
        // ------------------------------------------------------
        // Set properties and defaults
        // ------------------------------------------------------
        this.tags = props.tags;
        this.multiRegionProperties = props.multiRegionProperties;
        this.encryptionKey = props.kmsKey;
        // ------------------------------------------------------
        // Validation
        // ------------------------------------------------------
        this._validateMultiRegionProperties(this.multiRegionProperties);
        (0, validation_helpers_1.throwIfInvalid)(this._validateClusterTags, this.tags);
        this._validateRemovalPolicy(props.removalPolicy);
        // ------------------------------------------------------
        // CFN Props - With Lazy support
        // ------------------------------------------------------
        const cfnProps = {
            deletionProtectionEnabled: props.removalPolicy ? props.removalPolicy === aws_cdk_lib_1.RemovalPolicy.RETAIN : false,
            kmsEncryptionKey: this.encryptionKey?.keyArn,
            multiRegionProperties: this._renderMultiRegionProperties(),
            tags: this.tags ? Object.entries(this.tags).map(([key, value]) => ({ key, value })) : undefined,
        };
        // ------------------------------------------------------
        // Create the resource
        // ------------------------------------------------------
        this.__resource = new dsql.CfnCluster(this, 'Resource', cfnProps);
        // Get attributes directly from the CloudFormation resource
        this.clusterArn = this.__resource.attrResourceArn;
        this.creationTime = this.__resource.attrCreationTime;
        this.clusterId = this.__resource.attrIdentifier;
        this.status = this.__resource.attrStatus;
        this.vpcEndpointServiceName = this.__resource.attrVpcEndpointServiceName;
    }
    /**
     * Render the multi-region properties.
     *
     * @returns MultiRegionPropertiesProperty object in CloudFormation format, or undefined if no multi-region properties are defined
     * @default - undefined if no multi-region properties are provided
     * @internal This is an internal core function and should not be called directly.
     */
    _renderMultiRegionProperties() {
        return this.multiRegionProperties ? {
            clusters: this.multiRegionProperties.clusters?.map(cluster => cluster.clusterArn),
            witnessRegion: this.multiRegionProperties.witnessRegion,
        } : undefined;
    }
}
exports.Cluster = Cluster;
_b = JSII_RTTI_SYMBOL_1;
Cluster[_b] = { fqn: "@cdklabs/generative-ai-cdk-constructs.auroraDsql.Cluster", version: "0.1.312" };
