// @aws-cdk/aws-bedrock-alpha/bedrock/inference-profiles/application-inference-profile.js
"use strict";
var __runInitializers = (this && this.__runInitializers) || function (thisArg, initializers, value) {
    var useValue = arguments.length > 2;
    for (var i = 0; i < initializers.length; i++) {
        value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
    }
    return useValue ? value : void 0;
};
var __esDecorate = (this && this.__esDecorate) || function (ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
    function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
    var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
    var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
    var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
    var _, done = false;
    for (var i = decorators.length - 1; i >= 0; i--) {
        var context = {};
        for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
        for (var p in contextIn.access) context.access[p] = contextIn.access[p];
        context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
        var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
        if (kind === "accessor") {
            if (result === void 0) continue;
            if (result === null || typeof result !== "object") throw new TypeError("Object expected");
            if (_ = accept(result.get)) descriptor.get = _;
            if (_ = accept(result.set)) descriptor.set = _;
            if (_ = accept(result.init)) initializers.unshift(_);
        }
        else if (_ = accept(result)) {
            if (kind === "field") initializers.unshift(_);
            else descriptor[key] = _;
        }
    }
    if (target) Object.defineProperty(target, contextIn.name, descriptor);
    done = true;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApplicationInferenceProfile = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_cdk_lib_1 = require("aws-cdk-lib");
const bedrock = require("aws-cdk-lib/aws-bedrock");
const aws_iam_1 = require("aws-cdk-lib/aws-iam");
const metadata_resource_1 = require("aws-cdk-lib/core/lib/metadata-resource");
const prop_injectable_1 = require("aws-cdk-lib/core/lib/prop-injectable");
const inference_profile_1 = require("./inference-profile");
/******************************************************************************
 *                        NEW CONSTRUCT DEFINITION
 *****************************************************************************/
/**
 * Class to create an Application Inference Profile with CDK.
 * These are inference profiles created by users (user defined).
 * This helps to track costs and model usage.
 *
 * Application inference profiles are user-defined profiles that help you track costs and model usage.
 * They can be created for a single region or for multiple regions using a cross-region inference profile.
 *
 * @cloudformationResource AWS::Bedrock::ApplicationInferenceProfile
 * @see https://docs.aws.amazon.com/bedrock/latest/userguide/inference-profiles-create.html
 */
let ApplicationInferenceProfile = (() => {
    let _classDecorators = [prop_injectable_1.propertyInjectable];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    let _classSuper = inference_profile_1.InferenceProfileBase;
    let _instanceExtraInitializers = [];
    let _grantInvoke_decorators;
    let _grantProfileUsage_decorators;
    var ApplicationInferenceProfile = class extends _classSuper {
        static { _classThis = this; }
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _grantInvoke_decorators = [(0, metadata_resource_1.MethodMetadata)()];
            _grantProfileUsage_decorators = [(0, metadata_resource_1.MethodMetadata)()];
            __esDecorate(this, null, _grantInvoke_decorators, { kind: "method", name: "grantInvoke", static: false, private: false, access: { has: obj => "grantInvoke" in obj, get: obj => obj.grantInvoke }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(this, null, _grantProfileUsage_decorators, { kind: "method", name: "grantProfileUsage", static: false, private: false, access: { has: obj => "grantProfileUsage" in obj, get: obj => obj.grantProfileUsage }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
            ApplicationInferenceProfile = _classThis = _classDescriptor.value;
            if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        }
        static [JSII_RTTI_SYMBOL_1] = { fqn: "@aws-cdk/aws-bedrock-alpha.ApplicationInferenceProfile", version: "2.235.0-alpha.0" };
        /** Uniquely identifies this class. */
        static PROPERTY_INJECTION_ID = '@aws-cdk.aws-bedrock-alpha.ApplicationInferenceProfile';
        /**
         * Import an Application Inference Profile given its attributes.
         * [disable-awslint:no-grants]
         *
         * @param scope - The construct scope
         * @param id - Identifier of the construct
         * @param attrs - Attributes of the existing application inference profile
         * @returns An IInferenceProfile reference to the existing application inference profile
         */
        static fromApplicationInferenceProfileAttributes(scope, id, attrs) {
            try {
                jsiiDeprecationWarnings._aws_cdk_aws_bedrock_alpha_ApplicationInferenceProfileAttributes(attrs);
            }
            catch (error) {
                if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                    Error.captureStackTrace(error, this.fromApplicationInferenceProfileAttributes);
                }
                throw error;
            }
            class Import extends inference_profile_1.InferenceProfileBase {
                inferenceProfileArn = attrs.inferenceProfileArn;
                inferenceProfileId = aws_cdk_lib_1.Arn.split(attrs.inferenceProfileArn, aws_cdk_lib_1.ArnFormat.SLASH_RESOURCE_NAME)
                    .resourceName;
                type = inference_profile_1.InferenceProfileType.APPLICATION;
                /**
                 * [disable-awslint:no-grants]
                 */
                grantProfileUsage(grantee) {
                    return aws_iam_1.Grant.addToPrincipal({
                        grantee: grantee,
                        actions: ['bedrock:GetInferenceProfile', 'bedrock:InvokeModel'],
                        resourceArns: [this.inferenceProfileArn],
                    });
                }
            }
            return new Import(scope, id);
        }
        /**
         * Import a low-level L1 Cfn Application Inference Profile.
         * [disable-awslint:no-grants]
         *
         * @param cfnApplicationInferenceProfile - The L1 CfnApplicationInferenceProfile to import
         * @returns An IInferenceProfile reference to the imported application inference profile
         */
        static fromCfnApplicationInferenceProfile(cfnApplicationInferenceProfile) {
            // Singleton pattern that will prevent creating 2 or more Inference Profiles from the same L1
            const id = '@FromCfnApplicationInferenceProfile';
            const existing = cfnApplicationInferenceProfile.node.tryFindChild(id);
            if (existing) {
                return existing;
            }
            return new (class extends inference_profile_1.InferenceProfileBase {
                inferenceProfileArn = cfnApplicationInferenceProfile.attrInferenceProfileArn;
                inferenceProfileId = cfnApplicationInferenceProfile.attrInferenceProfileId;
                type = inference_profile_1.InferenceProfileType.APPLICATION;
                /**
                 * [disable-awslint:no-grants]
                 */
                grantProfileUsage(grantee) {
                    return aws_iam_1.Grant.addToPrincipal({
                        grantee: grantee,
                        actions: ['bedrock:GetInferenceProfile', 'bedrock:InvokeModel'],
                        resourceArns: [this.inferenceProfileArn],
                    });
                }
            })(cfnApplicationInferenceProfile, id);
        }
        // ------------------------------------------------------
        // Base attributes
        // ------------------------------------------------------
        /**
         * The name of the application inference profile.
         */
        inferenceProfileName = __runInitializers(this, _instanceExtraInitializers);
        /**
         * The ARN of the application inference profile.
         * @attribute
         */
        inferenceProfileArn;
        /**
         * The unique identifier of the application inference profile.
         * @attribute
         */
        inferenceProfileId;
        /**
         * The underlying model/cross-region model used by the application inference profile.
         */
        inferenceProfileModel;
        /**
         * The status of the application inference profile. ACTIVE means that the inference profile is ready to be used.
         * @attribute
         */
        status;
        /**
         * The type of the inference profile. Always APPLICATION for application inference profiles.
         */
        type;
        /**
         * Time Stamp for Application Inference Profile creation.
         * @attribute
         */
        createdAt;
        /**
         * Time Stamp for Application Inference Profile update.
         * @attribute
         */
        updatedAt;
        /**
         * The ARN used for invoking this inference profile.
         * This equals to the inferenceProfileArn property, useful for implementing IBedrockInvokable interface.
         */
        invokableArn;
        // ------------------------------------------------------
        // Internal Only
        // ------------------------------------------------------
        /**
         * Instance of CfnApplicationInferenceProfile.
         * @internal
         */
        __resource;
        // ------------------------------------------------------
        // CONSTRUCTOR
        // ------------------------------------------------------
        constructor(scope, id, props) {
            super(scope, id);
            try {
                jsiiDeprecationWarnings._aws_cdk_aws_bedrock_alpha_ApplicationInferenceProfileProps(props);
            }
            catch (error) {
                if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                    Error.captureStackTrace(error, ApplicationInferenceProfile);
                }
                throw error;
            }
            // Enhanced CDK Analytics Telemetry
            (0, metadata_resource_1.addConstructMetadata)(this, props);
            // ------------------------------------------------------
            // Validate props
            // ------------------------------------------------------
            this.validateProps(props);
            // ------------------------------------------------------
            // Set properties
            // ------------------------------------------------------
            this.inferenceProfileModel = props.modelSource;
            this.type = inference_profile_1.InferenceProfileType.APPLICATION;
            // ------------------------------------------------------
            // L1 instantiation
            // ------------------------------------------------------
            this.__resource = new bedrock.CfnApplicationInferenceProfile(this, 'Resource', {
                description: props.description,
                inferenceProfileName: props.applicationInferenceProfileName,
                modelSource: {
                    copyFrom: props.modelSource.invokableArn,
                },
                tags: props.tags ? Object.entries(props.tags).map(([key, value]) => ({ key, value })) : undefined,
            });
            // ------------------------------------------------------
            // Build attributes
            // ------------------------------------------------------
            this.inferenceProfileArn = this.__resource.attrInferenceProfileArn;
            this.inferenceProfileId = this.__resource.attrInferenceProfileId;
            this.inferenceProfileName = this.__resource.inferenceProfileName;
            this.status = this.__resource.attrStatus;
            this.createdAt = this.__resource.attrCreatedAt;
            this.updatedAt = this.__resource.attrUpdatedAt;
            this.invokableArn = this.inferenceProfileArn;
        }
        // ------------------------------------------------------
        // METHODS
        // ------------------------------------------------------
        /**
         * Validates the properties for creating an Application Inference Profile.
         *
         * @param props - The properties to validate
         * @throws ValidationError if any validation fails
         */
        validateProps(props) {
            // Validate applicationInferenceProfileName is provided and not empty
            if (!props.applicationInferenceProfileName || props.applicationInferenceProfileName.trim() === '') {
                throw new aws_cdk_lib_1.ValidationError('applicationInferenceProfileName is required and cannot be empty', this);
            }
            // Validate applicationInferenceProfileName length
            if (props.applicationInferenceProfileName.length > 64) {
                throw new aws_cdk_lib_1.ValidationError('applicationInferenceProfileName cannot exceed 64 characters', this);
            }
            // Validate applicationInferenceProfileName pattern
            const namePattern = /^([0-9a-zA-Z:.][ _-]?)+$/;
            if (!namePattern.test(props.applicationInferenceProfileName)) {
                throw new aws_cdk_lib_1.ValidationError('applicationInferenceProfileName must match pattern ^([0-9a-zA-Z:.][ _-]?)+$', this);
            }
            // Validate modelSource is provided
            if (!props.modelSource) {
                throw new aws_cdk_lib_1.ValidationError('modelSource is required', this);
            }
            // Validate description length if provided
            if (props.description !== undefined && props.description.length > 200) {
                throw new aws_cdk_lib_1.ValidationError('description cannot exceed 200 characters', this);
            }
            // Validate description pattern if provided
            if (props.description !== undefined && props.description !== '') {
                const descriptionPattern = /^([0-9a-zA-Z:.][ _-]?)+$/;
                if (!descriptionPattern.test(props.description)) {
                    throw new aws_cdk_lib_1.ValidationError('description must match pattern ^([0-9a-zA-Z:.][ _-]?)+$', this);
                }
            }
        }
        /**
         * Gives the appropriate policies to invoke and use the application inference profile.
         * This method ensures the appropriate permissions are given to use either the inference profile
         * or the underlying foundation model/cross-region profile.
         * [disable-awslint:no-grants]
         *
         * @param grantee - The IAM principal to grant permissions to
         * @returns An IAM Grant object representing the granted permissions
         */
        grantInvoke(grantee) {
            // This method ensures the appropriate permissions are given
            // to use either the inference profile or the vanilla foundation model
            this.inferenceProfileModel.grantInvoke(grantee);
            // Plus we add permissions to now invoke the application inference profile itself
            return this.grantProfileUsage(grantee);
        }
        /**
         * Grants appropriate permissions to use the application inference profile (AIP).
         * This method adds the necessary IAM permissions to allow the grantee to:
         * - Get inference profile details (bedrock:GetInferenceProfile)
         * - Invoke the model through the inference profile (bedrock:InvokeModel)
         *
         * Note: This does not grant permissions to use the underlying model/cross-region profile in the AIP.
         * For comprehensive permissions, use grantInvoke() instead.
         * [disable-awslint:no-grants]
         *
         * @param grantee - The IAM principal to grant permissions to
         * @returns An IAM Grant object representing the granted permissions
         */
        grantProfileUsage(grantee) {
            return aws_iam_1.Grant.addToPrincipal({
                grantee: grantee,
                actions: ['bedrock:GetInferenceProfile', 'bedrock:InvokeModel'],
                resourceArns: [this.inferenceProfileArn],
            });
        }
        static {
            __runInitializers(_classThis, _classExtraInitializers);
        }
    };
    return ApplicationInferenceProfile = _classThis;
})();
exports.ApplicationInferenceProfile = ApplicationInferenceProfile;
