// @aws-cdk/aws-bedrock-alpha/bedrock/inference-profiles/cross-region-inference-profile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CrossRegionInferenceProfile = exports.REGION_TO_GEO_AREA = exports.CrossRegionInferenceProfileRegion = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_cdk_lib_1 = require("aws-cdk-lib");
const aws_iam_1 = require("aws-cdk-lib/aws-iam");
const inference_profile_1 = require("./inference-profile");
/**
 * Error thrown when cross-region inference profile validation fails.
 */
class CrossRegionInferenceProfileError extends Error {
    constructor(message) {
        super(message);
        this.name = 'CrossRegionInferenceProfileError';
    }
}
/**
 * Geographic regions supported for cross-region inference profiles.
 * These regions help distribute traffic across multiple AWS regions for better
 * throughput and resilience during peak demands.
 */
var CrossRegionInferenceProfileRegion;
(function (CrossRegionInferenceProfileRegion) {
    /**
     * Global cross-region Inference Identifier.
     * Routes requests to any supported commercial AWS Region.
     */
    CrossRegionInferenceProfileRegion["GLOBAL"] = "global";
    /**
     * Cross-region Inference Identifier for the European area.
     * According to the model chosen, this might include:
     * - Frankfurt (`eu-central-1`)
     * - Ireland (`eu-west-1`)
     * - Paris (`eu-west-3`)
     * - London (`eu-west-2`)
     * - Stockholm (`eu-north-1`)
     * - Milan (`eu-south-1`)
     * - Spain (`eu-south-2`)
     * - Zurich (`eu-central-2`)
     */
    CrossRegionInferenceProfileRegion["EU"] = "eu";
    /**
     * Cross-region Inference Identifier for the United States area.
     * According to the model chosen, this might include:
     * - N. Virginia (`us-east-1`)
     * - Ohio (`us-east-2`)
     * - Oregon (`us-west-2`)
     */
    CrossRegionInferenceProfileRegion["US"] = "us";
    /**
     * Cross-region Inference Identifier for the US GovCloud area.
     * According to the model chosen, this might include:
     * - GovCloud US-East (`us-gov-east-1`)
     * - GovCloud US-West (`us-gov-west-1`)
     */
    CrossRegionInferenceProfileRegion["US_GOV"] = "us-gov";
    /**
     * Cross-region Inference Identifier for the Asia-Pacific area.
     * According to the model chosen, this might include:
     * - Tokyo (`ap-northeast-1`)
     * - Seoul (`ap-northeast-2`)
     * - Osaka (`ap-northeast-3`)
     * - Mumbai (`ap-south-1`)
     * - Hyderabad (`ap-south-2`)
     * - Singapore (`ap-southeast-1`)
     * - Sydney (`ap-southeast-2`)
     * - Jakarta (`ap-southeast-3`)
     * - Melbourne (`ap-southeast-4`)
     * - Malaysia (`ap-southeast-5`)
     * - Thailand (`ap-southeast-7`)
     * - Taipei (`ap-east-2`)
     * - Middle East (UAE) (`me-central-1`)
     */
    CrossRegionInferenceProfileRegion["APAC"] = "apac";
    /**
     * Cross-region Inference Identifier for the Japan area.
     * According to the model chosen, this might include:
     * - Tokyo (`ap-northeast-1`)
     * - Osaka (`ap-northeast-3`)
     */
    CrossRegionInferenceProfileRegion["JP"] = "jp";
    /**
     * Cross-region Inference Identifier for the Australia area.
     * According to the model chosen, this might include:
     * - Sydney (`ap-southeast-2`)
     * - Melbourne (`ap-southeast-4`)
     */
    CrossRegionInferenceProfileRegion["AU"] = "au";
})(CrossRegionInferenceProfileRegion || (exports.CrossRegionInferenceProfileRegion = CrossRegionInferenceProfileRegion = {}));
/**
 * Mapping of AWS regions to their corresponding geographic areas for cross-region inference.
 * This mapping is used to determine which cross-region inference profile to use based on the current region in prompt router.
 */
exports.REGION_TO_GEO_AREA = {
    // US Regions
    'us-east-1': CrossRegionInferenceProfileRegion.US, // N. Virginia
    'us-east-2': CrossRegionInferenceProfileRegion.US, // Ohio
    'us-west-2': CrossRegionInferenceProfileRegion.US, // Oregon
    // EU Regions
    'eu-central-1': CrossRegionInferenceProfileRegion.EU, // Frankfurt
    'eu-west-1': CrossRegionInferenceProfileRegion.EU, // Ireland
    'eu-west-3': CrossRegionInferenceProfileRegion.EU, // Paris
    // APAC Regions
    'ap-northeast-1': CrossRegionInferenceProfileRegion.APAC, // Tokyo
    'ap-northeast-2': CrossRegionInferenceProfileRegion.APAC, // Seoul
    'ap-south-1': CrossRegionInferenceProfileRegion.APAC, // Mumbai
    'ap-southeast-1': CrossRegionInferenceProfileRegion.APAC, // Singapore
    'ap-southeast-2': CrossRegionInferenceProfileRegion.APAC, // Sydney
};
/******************************************************************************
 *                        NEW CONSTRUCT DEFINITION
 *****************************************************************************/
/**
 * Cross-region inference enables you to seamlessly manage unplanned traffic
 * bursts by utilizing compute across different AWS Regions. With cross-region
 * inference, you can distribute traffic across multiple AWS Regions, enabling
 * higher throughput and enhanced resilience during periods of peak demands.
 *
 * This construct represents a system-defined inference profile that routes
 * requests across multiple regions based on availability and demand.
 *
 * @see https://docs.aws.amazon.com/bedrock/latest/userguide/cross-region-inference.html
 */
class CrossRegionInferenceProfile {
    static [JSII_RTTI_SYMBOL_1] = { fqn: "@aws-cdk/aws-bedrock-alpha.CrossRegionInferenceProfile", version: "2.235.0-alpha.0" };
    /**
     * Creates a Cross-Region Inference Profile from the provided configuration.
     *
     * @param config - Configuration for the cross-region inference profile
     * @returns A new CrossRegionInferenceProfile instance
     * @throws ValidationError if the model doesn't support cross-region inference
     */
    static fromConfig(config) {
        try {
            jsiiDeprecationWarnings._aws_cdk_aws_bedrock_alpha_CrossRegionInferenceProfileProps(config);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.fromConfig);
            }
            throw error;
        }
        return new CrossRegionInferenceProfile(config);
    }
    /**
     * The unique identifier of the inference profile.
     * Format: {geoRegion}.{modelId}
     */
    inferenceProfileId;
    /**
     * The ARN of the inference profile.
     * @attribute
     */
    inferenceProfileArn;
    /**
     * The type of inference profile. Always SYSTEM_DEFINED for cross-region profiles.
     */
    type;
    /**
     * The underlying foundation model supporting cross-region inference.
     */
    inferenceProfileModel;
    /**
     * The ARN used for invoking this inference profile.
     * This equals to the inferenceProfileArn property, useful for implementing IBedrockInvokable interface.
     */
    invokableArn;
    constructor(props) {
        // Validate required properties
        if (!props.geoRegion) {
            throw new CrossRegionInferenceProfileError('geoRegion is required');
        }
        if (!props.model) {
            throw new CrossRegionInferenceProfileError('model is required');
        }
        // Validate that the model supports cross-region inference
        if (!props.model.supportsCrossRegion) {
            throw new CrossRegionInferenceProfileError(`Model ${props.model.modelId} does not support cross-region inference`);
        }
        this.type = inference_profile_1.InferenceProfileType.SYSTEM_DEFINED;
        this.inferenceProfileModel = props.model;
        this.inferenceProfileId = `${props.geoRegion}.${props.model.modelId}`;
        this.inferenceProfileArn = aws_cdk_lib_1.Arn.format({
            partition: aws_cdk_lib_1.Aws.PARTITION,
            service: 'bedrock',
            account: aws_cdk_lib_1.Aws.ACCOUNT_ID,
            region: aws_cdk_lib_1.Aws.REGION,
            resource: 'inference-profile',
            resourceName: this.inferenceProfileId,
            arnFormat: aws_cdk_lib_1.ArnFormat.SLASH_RESOURCE_NAME,
        });
        this.invokableArn = this.inferenceProfileArn;
    }
    /**
     * Gives the appropriate policies to invoke and use the Foundation Model.
     * For cross-region inference profiles, this method grants permissions to:
     * - Invoke the model in all regions where the inference profile can route requests
     * - Use the inference profile itself
     * [disable-awslint:no-grants]
     *
     * @param grantee - The IAM principal to grant permissions to
     * @returns An IAM Grant object representing the granted permissions
     */
    grantInvoke(grantee) {
        // For cross-region inference profiles, we need to provide permissions to invoke the model in all regions
        // where the inference profile can route requests
        this.inferenceProfileModel.grantInvokeAllRegions(grantee);
        // And we need to provide permissions to invoke the inference profile itself
        return this.grantProfileUsage(grantee);
    }
    /**
     * Grants appropriate permissions to use the cross-region inference profile.
     * This method adds the necessary IAM permissions to allow the grantee to:
     * - Get inference profile details (bedrock:GetInferenceProfile)
     * - Invoke the model through the inference profile (bedrock:InvokeModel*)
     *
     * Note: This does not grant permissions to use the underlying model directly.
     * For comprehensive permissions, use grantInvoke() instead.
     * [disable-awslint:no-grants]
     *
     * @param grantee - The IAM principal to grant permissions to
     * @returns An IAM Grant object representing the granted permissions
     */
    grantProfileUsage(grantee) {
        return aws_iam_1.Grant.addToPrincipal({
            grantee: grantee,
            actions: ['bedrock:GetInferenceProfile', 'bedrock:InvokeModel*'],
            resourceArns: [this.inferenceProfileArn],
        });
    }
}
exports.CrossRegionInferenceProfile = CrossRegionInferenceProfile;
