// @aws-cdk/aws-bedrock-alpha/bedrock/inference-profiles/prompt-router.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PromptRouter = exports.DefaultPromptRouterIdentifier = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_cdk_lib_1 = require("aws-cdk-lib");
const aws_iam_1 = require("aws-cdk-lib/aws-iam");
const models_1 = require("../models");
const cross_region_inference_profile_1 = require("./cross-region-inference-profile");
/**
 * Represents identifiers for default prompt routers in Bedrock.
 * These are pre-configured routers provided by AWS that route between
 * different models in the same family for optimal performance and cost.
 */
class DefaultPromptRouterIdentifier {
    static [JSII_RTTI_SYMBOL_1] = { fqn: "@aws-cdk/aws-bedrock-alpha.DefaultPromptRouterIdentifier", version: "2.235.0-alpha.0" };
    /**
     * Anthropic Claude V1 router configuration.
     * Routes between Claude Haiku and Claude 3.5 Sonnet models for optimal
     * balance between performance and cost.
     */
    static ANTHROPIC_CLAUDE_V1 = new DefaultPromptRouterIdentifier({
        promptRouterId: 'anthropic.claude:1',
        routingModels: [
            models_1.BedrockFoundationModel.ANTHROPIC_CLAUDE_HAIKU_V1_0,
            models_1.BedrockFoundationModel.ANTHROPIC_CLAUDE_3_5_SONNET_V1_0,
        ],
    });
    /**
     * Meta Llama 3.1 router configuration.
     * Routes between different sizes of Llama 3.1 models (8B and 70B)
     * for optimal performance based on request complexity.
     */
    static META_LLAMA_3_1 = new DefaultPromptRouterIdentifier({
        promptRouterId: 'meta.llama:1',
        routingModels: [
            models_1.BedrockFoundationModel.META_LLAMA_3_1_8B_INSTRUCT_V1,
            models_1.BedrockFoundationModel.META_LLAMA_3_1_70B_INSTRUCT_V1,
        ],
    });
    /**
     * The unique identifier for this prompt router.
     */
    promptRouterId;
    /**
     * The foundation models that this router can route between.
     */
    routingModels;
    constructor(props) {
        this.promptRouterId = props.promptRouterId;
        this.routingModels = props.routingModels;
    }
}
exports.DefaultPromptRouterIdentifier = DefaultPromptRouterIdentifier;
/**
 * Amazon Bedrock intelligent prompt routing provides a single serverless endpoint
 * for efficiently routing requests between different foundational models within
 * the same model family. It can help you optimize for response quality and cost.
 *
 * Intelligent prompt routing predicts the performance of each model for each request,
 * and dynamically routes each request to the model that it predicts is most likely
 * to give the desired response at the lowest cost.
 *
 * @see https://docs.aws.amazon.com/bedrock/latest/userguide/prompt-routing.html
 */
class PromptRouter {
    static [JSII_RTTI_SYMBOL_1] = { fqn: "@aws-cdk/aws-bedrock-alpha.PromptRouter", version: "2.235.0-alpha.0" };
    /**
     * Creates a PromptRouter from a default router identifier.
     *
     * @param defaultRouter - The default router configuration to use
     * @param region - The AWS region where the router will be used
     * @returns A new PromptRouter instance configured with the default settings
     */
    static fromDefaultId(defaultRouter, region) {
        try {
            jsiiDeprecationWarnings._aws_cdk_aws_bedrock_alpha_DefaultPromptRouterIdentifier(defaultRouter);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.fromDefaultId);
            }
            throw error;
        }
        return new PromptRouter(defaultRouter, region);
    }
    /**
     * The ARN of the prompt router.
     * @attribute
     */
    promptRouterArn;
    /**
     * The ID of the prompt router.
     * @attribute
     */
    promptRouterId;
    /**
     * The ARN used for invoking this prompt router.
     * This equals to the promptRouterArn property, useful for implementing IBedrockInvokable interface.
     */
    invokableArn;
    /**
     * The inference endpoints (cross-region profiles) that this router will route to.
     * These are created automatically based on the routing models and region.
     */
    routingEndpoints;
    constructor(props, region) {
        try {
            jsiiDeprecationWarnings._aws_cdk_aws_bedrock_alpha_PromptRouterProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, PromptRouter);
            }
            throw error;
        }
        this.promptRouterId = props.promptRouterId;
        this.promptRouterArn = aws_cdk_lib_1.Arn.format({
            partition: aws_cdk_lib_1.Aws.PARTITION,
            service: 'bedrock',
            region: region,
            account: aws_cdk_lib_1.Aws.ACCOUNT_ID,
            resource: 'default-prompt-router',
            resourceName: this.promptRouterId,
            arnFormat: aws_cdk_lib_1.ArnFormat.SLASH_RESOURCE_NAME,
        });
        // Needed to implement IBedrockInvokable
        this.invokableArn = this.promptRouterArn;
        // Build inference profiles from routing endpoints
        // Each routing model is wrapped in a cross-region inference profile
        // to enable routing across regions for better availability
        this.routingEndpoints = props.routingModels.flatMap(model => {
            const geoRegion = cross_region_inference_profile_1.REGION_TO_GEO_AREA[region];
            if (geoRegion) {
                return cross_region_inference_profile_1.CrossRegionInferenceProfile.fromConfig({
                    model: model,
                    geoRegion: geoRegion,
                });
            }
            else {
                // For unknown regions, fall back to using the model directly
                return model;
            }
        });
    }
    /**
     * Grants the necessary permissions to invoke this prompt router and all its routing endpoints.
     * This method grants permissions to:
     * - Get prompt router details (bedrock:GetPromptRouter)
     * - Invoke models through the router (bedrock:InvokeModel)
     * - Use all underlying models and cross-region profiles
     * [disable-awslint:no-grants]
     *
     * @param grantee - The IAM principal to grant permissions to
     * @returns An IAM Grant object representing the granted permissions
     */
    grantInvoke(grantee) {
        // Grant invoke permissions on every model endpoint of the router
        this.routingEndpoints.forEach(model => {
            model.grantInvoke(grantee);
        });
        // Grant invoke permissions to the prompt router itself
        return aws_iam_1.Grant.addToPrincipal({
            grantee,
            actions: ['bedrock:GetPromptRouter', 'bedrock:InvokeModel'],
            resourceArns: [this.promptRouterArn],
        });
    }
}
exports.PromptRouter = PromptRouter;
