// awslint/lib/linter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DiagnosticLevel = exports.Evaluation = exports.Linter = exports.AggregateLinter = exports.LinterBase = void 0;
const util = require("util");
const reflect = require("jsii-reflect");
class LinterBase {
}
exports.LinterBase = LinterBase;
class AggregateLinter extends LinterBase {
    linters;
    constructor(...linters) {
        super();
        this.linters = linters;
    }
    get rules() {
        const ret = new Array();
        for (const linter of this.linters) {
            ret.push(...linter.rules);
        }
        return ret;
    }
    eval(assembly, options) {
        const diags = new Array();
        for (const linter of this.linters) {
            diags.push(...linter.eval(assembly, options));
        }
        return diags;
    }
}
exports.AggregateLinter = AggregateLinter;
/**
 * Evaluates a bunch of rules against some context.
 */
class Linter extends LinterBase {
    init;
    _rules = {};
    constructor(init) {
        super();
        this.init = init;
    }
    get rules() {
        return Object.values(this._rules);
    }
    /**
     * Install another rule.
     */
    add(rule) {
        if (rule.code in this._rules) {
            throw new Error(`rule "${rule.code}" already exists`);
        }
        this._rules[rule.code] = rule;
    }
    /**
     * Evaluate all rules against the context.
     */
    eval(assembly, options) {
        options = options || {};
        let ctxs = this.init(assembly);
        if (!ctxs) {
            return []; // skip
        }
        if (!Array.isArray(ctxs)) {
            ctxs = [ctxs];
        }
        const diag = new Array();
        for (const ctx of ctxs) {
            for (const rule of Object.values(this._rules)) {
                const evaluation = new Evaluation(ctx, rule, diag, options);
                rule.eval(evaluation);
            }
        }
        return diag;
    }
}
exports.Linter = Linter;
/**
 * Passed in to each rule during evaluation.
 */
class Evaluation {
    ctx;
    options;
    curr;
    diagnostics;
    constructor(ctx, rule, diagnostics, options) {
        this.ctx = ctx;
        this.options = options;
        this.curr = rule;
        this.diagnostics = diagnostics;
    }
    /**
     * Record a failure if `condition` is not truthy.
     *
     * @param condition The condition to assert.
     * @param scope Used to diagnose the location in the source, and is used in the
     * ignore pattern.
     * @param extra Used to replace %s in the default message format string.
     */
    assert(condition, scope, extra) {
        // deduplicate: skip if this specific assertion ("rule:scope") was already examined
        if (this.diagnostics.find(d => d.rule.code === this.curr.code && d.scope === scope)) {
            return condition;
        }
        const include = this.shouldEvaluate(this.curr.code, scope);
        const message = util.format(this.curr.message, extra || '');
        // Don't add a "Success" diagnostic. It will break if we run a compound
        // linter rule which consists of 3 checks with the same scope (such
        // as for example `assertSignature()`). If the first check fails, we would
        // add a "Success" diagnostic and all other diagnostics would be skipped because
        // of the deduplication check above. Changing the scope makes it worse, since
        // the scope is also the ignore pattern and they're all conceptually the same rule.
        //
        // Simplest solution is to not record successes -- why do we even need them?
        if (include && condition) {
            return condition;
        }
        let level;
        if (!include) {
            level = DiagnosticLevel.Skipped;
        }
        else if (this.curr.warning) {
            level = DiagnosticLevel.Warning;
        }
        else {
            level = DiagnosticLevel.Error;
        }
        const diag = {
            level,
            rule: this.curr,
            scope,
            message,
        };
        this.diagnostics.push(diag);
        return condition;
    }
    assertEquals(actual, expected, scope) {
        return this.assert(actual === expected, scope, ` (expected="${expected}",actual="${actual}")`);
    }
    assertTypesEqual(ts, actual, expected, scope) {
        const a = typeReferenceFrom(ts, actual);
        const e = typeReferenceFrom(ts, expected);
        return this.assert(a.toString() === e.toString(), scope, ` (expected="${e}",actual="${a}")`);
    }
    assertTypesAssignable(ts, actual, expected, scope) {
        const a = typeReferenceFrom(ts, actual);
        const e = typeReferenceFrom(ts, expected);
        return this.assert(a.toString() === e.toString() || (a.fqn && e.fqn && a.type.extends(e.type)), scope, ` ("${a}" not assignable to "${e}")`);
    }
    assertParameterOptional(actual, expected, scope) {
        return this.assert(actual === expected, scope, ` (${scope} should be ${expected ? 'optional' : 'mandatory'})`);
    }
    assertSignature(method, expectations) {
        const scope = method.parentType.fqn + '.' + method.name;
        if (expectations.returns && reflect.Method.isMethod(method)) {
            this.assertTypesEqual(method.system, method.returns.type, expectations.returns, scope);
        }
        if (expectations.parameters) {
            const expectedCount = expectations.parameters.length;
            const actualCount = method.parameters.length;
            if (this.assertEquals(actualCount, expectedCount, scope)) {
                for (let i = 0; i < expectations.parameters.length; ++i) {
                    const expect = expectations.parameters[i];
                    const actual = method.parameters[i];
                    const pscope = scope + `.params[${i}]`;
                    if (expect.name) {
                        const expectedName = expect.name;
                        const actualName = actual.name;
                        this.assertEquals(actualName, expectedName, pscope);
                    }
                    if (expect.type) {
                        if (expect.subtypeAllowed) {
                            this.assertTypesAssignable(method.system, actual.type, expect.type, pscope);
                        }
                        else {
                            this.assertTypesEqual(method.system, actual.type, expect.type, pscope);
                        }
                    }
                    if (expect.optional !== undefined) {
                        this.assertParameterOptional(actual.optional, expect.optional, pscope);
                    }
                }
            }
        }
    }
    /**
     * Evaluates whether the rule should be evaluated based on the filters applied.
     */
    shouldEvaluate(code, scope) {
        if (!this.options.includeRules || this.options.includeRules.isEmpty()) {
            return true;
        }
        if (this.options.includeRules.matches(code, scope)) {
            if (this.options.excludeRules?.matches(code, scope)) {
                return false;
            }
            return true;
        }
        return false;
    }
}
exports.Evaluation = Evaluation;
var DiagnosticLevel;
(function (DiagnosticLevel) {
    DiagnosticLevel[DiagnosticLevel["Skipped"] = 0] = "Skipped";
    DiagnosticLevel[DiagnosticLevel["Success"] = 1] = "Success";
    DiagnosticLevel[DiagnosticLevel["Warning"] = 2] = "Warning";
    DiagnosticLevel[DiagnosticLevel["Error"] = 3] = "Error";
})(DiagnosticLevel || (exports.DiagnosticLevel = DiagnosticLevel = {}));
/**
 * Convert a type specifier to a TypeReference
 */
function typeReferenceFrom(ts, x) {
    if (isTypeReference(x)) {
        return x;
    }
    if (typeof x === 'string') {
        if (x.indexOf('.') === -1) {
            return new reflect.TypeReference(ts, { primitive: x });
        }
        else {
            return new reflect.TypeReference(ts, { fqn: x });
        }
    }
    return new reflect.TypeReference(ts, x);
}
function isTypeReference(x) {
    return x instanceof reflect.TypeReference;
}
